{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare variant calls and annotations from two RNA-seq experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "545 unique variants in experiment two\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "#input_file1 = \"/cluster/db/mecoulter/CEMartinez_RNAseq-209915706/results/bcftools/3H_locus_filtered.anno.vcf\"\n",
    "input_file2 = \"/cluster/db/mecoulter/RNAseq2/bcftools/3H_locus_filtered.anno.vcf\"\n",
    "\n",
    "first_set = set()\n",
    "second_set = set()\n",
    "\n",
    "locus = [33181340, 36970860]#3H locus boundaries on 3H\n",
    "\"\"\"\n",
    "for line in open(input_file1):#Barke - Int19/Int56\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    fields = line.rstrip().split(\"\\t\")\n",
    "    SNP_position = fields[1]\n",
    "    ref = fields[3]\n",
    "    alt = fields[4]\n",
    "    variant_id = f\"{SNP_position}_{ref}_{alt}\"\n",
    "    first_set.add(variant_id)\n",
    "\n",
    "print(f\"{len(first_set)} unique variants in experiment one\")\n",
    "\"\"\"\n",
    "for line in open(input_file2):#Barke/Int_124_17 - Int_124_52\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    fields = line.rstrip().split(\"\\t\")\n",
    "    SNP_position = fields[1]\n",
    "    if int(SNP_position) < locus[0] or int(SNP_position) > locus[1]:\n",
    "        continue\n",
    "    ref = fields[3]\n",
    "    alt = fields[4]\n",
    "    variant_id = f\"{SNP_position}_{ref}_{alt}\"\n",
    "    second_set.add(variant_id)\n",
    "\n",
    "print(f\"{len(second_set)} unique variants in experiment two\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* What variants are shared between two experiments?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(f\"Number of shared variants: {len(first_set & second_set)}, proportion (compared to first set): {len(first_set & second_set)/len(first_set)} \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KeyError for gene BaRT2v18chr3HG122940-BaRT2v18chr3HG122950\n",
      "KeyError for gene BaRT2v18chr3HG122950-BaRT2v18chr3HG122960\n",
      "KeyError for gene BaRT2v18chr3HG122980-BaRT2v18chr3HG122990\n",
      "KeyError for gene BaRT2v18chr3HG122990-BaRT2v18chr3HG123000\n",
      "KeyError for gene BaRT2v18chr3HG123000-BaRT2v18chr3HG123010\n",
      "KeyError for gene BaRT2v18chr3HG123010-BaRT2v18chr3HG123030\n",
      "KeyError for gene BaRT2v18chr3HG123030-BaRT2v18chr3HG123040\n",
      "KeyError for gene BaRT2v18chr3HG123040-BaRT2v18chr3HG123060\n",
      "KeyError for gene BaRT2v18chr3HG123110-BaRT2v18chr3HG123120\n",
      "KeyError for gene BaRT2v18chr3HG123150-BaRT2v18chr3HG123170\n",
      "KeyError for gene BaRT2v18chr3HG123180-BaRT2v18chr3HG123200\n",
      "KeyError for gene BaRT2v18chr3HG123210-BaRT2v18chr3HG123220\n",
      "KeyError for gene BaRT2v18chr3HG123390-BaRT2v18chr3HG123400\n",
      "KeyError for gene BaRT2v18chr3HG123410-BaRT2v18chr3HG123420\n",
      "KeyError for gene BaRT2v18chr3HG123420-BaRT2v18chr3HG123430\n",
      "KeyError for gene BaRT2v18chr3HG123440-BaRT2v18chr3HG123450\n",
      "KeyError for gene BaRT2v18chr3HG123490-BaRT2v18chr3HG123500\n"
     ]
    }
   ],
   "source": [
    "input_file = input_file2\n",
    "\n",
    "annotation_input = \"/cluster/db/mecoulter/BaRT2v18/BaRT_2_18_annotation_genes.txt\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class SnpEffAnnotation:\n",
    "    def __init__(self, annotation):\n",
    "        try:\n",
    "            self.allele, self.p_annotation, self.putative_impact, self.position, self.gene, self.feature_type, self.transcript, self.transcript_type  = annotation.split(\"|\")[:8]\n",
    "        except ValueError:\n",
    "            print(f\"Error at position {SNP_position}! Annotation is {annotation}\")\n",
    "            sys.exit()\n",
    "    def __repr__(self):\n",
    "        return self.p_annotation\n",
    "\n",
    "class SnpEff:\n",
    "    gene_impact = {}\n",
    "    gene_SNP_annotation = {}\n",
    "    Instances = {}#Gene: {position1: [annotations], position2: [annotations]}\n",
    "    Instances_clean = {}\n",
    "    position_info = {}\n",
    "    dn_ds = {}\n",
    "    \n",
    "    def __init__(self, line):\n",
    "        self.fields = line.rstrip().split(\"\\t\")\n",
    "        self.INFO = self.fields[7]\n",
    "        self.SNP_position = self.fields[1]\n",
    "        self.reference_allele = self.fields[3]\n",
    "        self.alternative_allele = self.fields[4]\n",
    "        annotations = self.INFO.split(\"ANN=\")[1].split(\"=\")[0].split(\",\")#List of all possible annotations\n",
    "        self.annotations = []\n",
    "        for a in annotations:\n",
    "            annotation = SnpEffAnnotation(a)\n",
    "            self.annotations.append(annotation)\n",
    "            self.gene_impact.setdefault(annotation.gene,[]).append(annotation.putative_impact)  \n",
    "            \n",
    "            if annotation.putative_impact == \"HIGH\" or annotation.putative_impact == \"MODERATE\":\n",
    "                self.gene_SNP_annotation.setdefault(annotation.gene,[]).append(f\"{annotation.p_annotation} : {self.SNP_position}\")\n",
    "                \n",
    "            try:\n",
    "                position_annotations = self.Instances[annotation.gene]\n",
    "                position_annotations.setdefault(self.SNP_position, []).append(annotation)\n",
    "                self.Instances[annotation.gene] = position_annotations\n",
    "            except KeyError:\n",
    "                positions_annotations = {}\n",
    "                positions_annotations[self.SNP_position] = [annotation]\n",
    "                self.Instances[annotation.gene] = positions_annotations\n",
    "                \n",
    "        self.position_info[self.SNP_position] = self\n",
    "                \n",
    "    def __getitem__(self, item):\n",
    "        if self.Instances_clean:\n",
    "            return self.Instances_clean[item]\n",
    "        else:\n",
    "            return self.Instances[item]\n",
    "    \n",
    "    def clean_up_annotations(self):\n",
    "        \"\"\"Many positions have multiple annotations.\n",
    "        This is because of multiple transcripts. Most transcripts will be similar, \n",
    "        so will effectively have the same annotation. So remove duplicates. Will\n",
    "        also remove annotations not of interest e.g upstream/ downstream\"\"\"\n",
    "        \n",
    "        not_interesting = {\"downstream_gene_variant\", \"upstream_gene_variant\", \\\n",
    "        \"intergenic_region\", \"intron_variant\"}\n",
    "        \n",
    "        for gene, position_annotation in self.Instances.items():\n",
    "            position_annotation_clean = {}\n",
    "            for position, annotations in position_annotation.items():\n",
    "                annotations_cleaned = []\n",
    "                all_p_annotations = set()\n",
    "                for annotation in annotations:\n",
    "                    if annotation.p_annotation not in all_p_annotations and annotation.p_annotation not in not_interesting:\n",
    "                        #if annotation.putative_impact == \"HIGH\" or annotation.putative_impact == \"MODERATE\":\n",
    "                        annotations_cleaned.append(annotation)\n",
    "                        all_p_annotations.add(annotation.p_annotation)\n",
    "                if annotations_cleaned:#Only keep high/moderate impact snps\n",
    "                    position_annotation_clean[position] = annotations_cleaned\n",
    "            self.Instances_clean[gene] = position_annotation_clean\n",
    "    def __repr__(self):\n",
    "        string = \"\"\n",
    "        print_dict = self.Instances_clean if self.Instances_clean else self.Instances\n",
    "        \n",
    "        for gene, position_annotation in print_dict.items():\n",
    "            string += f\"{gene}: \\n\"\n",
    "            for position, annotations in position_annotation.items():\n",
    "                string += f\"{position} : {','.join([str(annotation) for annotation in annotations])}\\n\"\n",
    "            string += \"\\n\"  \n",
    "        return string\n",
    "    \n",
    "    def __len__(self):\n",
    "        return(len(self.Instances.keys()))\n",
    "    \n",
    "    def write_tsv(self, outfile, gene_annotation):\n",
    "        \"\"\"Write a .csv file with results of cleaned up annotation. \n",
    "        Gene: Position: Annotation\"\"\"\n",
    "        with open(outfile, \"w\") as out:\n",
    "            out.write(\"Gene\\tPosition\\tAnnotation\\tref allele\\talt allele\\tPanzzer Annotation\\tdn/ds ratio\\n\")\n",
    "            for gene, position_annotation in self.Instances_clean.items():\n",
    "                \n",
    "                index = 0\n",
    "                \n",
    "                try:\n",
    "                    panz_annotation = gene_annotation.at[gene,\"Pannzer annotation\"]\n",
    "                except KeyError:\n",
    "                    print(f\"KeyError for gene {gene}\")\n",
    "                    panz_annotation = \"Not found\"\n",
    "                    \n",
    "                try:\n",
    "                    dn_ds = self.dn_ds[gene]\n",
    "                except KeyError:\n",
    "                    dn_ds = \"\"\n",
    "                    \n",
    "                for position, annotations in position_annotation.items():\n",
    "                    \n",
    "                    ref = self.position_info[position].reference_allele\n",
    "                    alt = self.position_info[position].alternative_allele\n",
    "                    \n",
    "                    out.write(f\"{gene}\\t\") if not index else out.write(f\"\\t\")\n",
    "                    \n",
    "                    out.write(f\"{position}\\t{','.join([str(annotation) for annotation in annotations])}\\t\\\n",
    "                    {ref}\\t{alt}\\t\")#\n",
    "                    \n",
    "                    #Only write annotation once\n",
    "                    out.write(f\"{panz_annotation}\\t{dn_ds}\\n\") if not index else out.write(\"\\t\\n\")\n",
    "                    \n",
    "                    index += 1\n",
    "            \n",
    "    def get_dn_ds(self):\n",
    "        \n",
    "        \"\"\"Calculates dn/ds ratio for each gene where possible\n",
    "        missense_variant = n, synonymous_variant\"\"\"\n",
    "        for gene, position_annotation in self.Instances_clean.items():\n",
    "            n = []\n",
    "            s = []\n",
    "            for position, annotations in position_annotation.items():\n",
    "                SNP_anotations = set([str(annotation) for annotation in annotations])\n",
    "                if \"missense_variant\" in SNP_anotations and \"synonymous_variant\" in SNP_anotations:\n",
    "                    continue#A case where SNP is annotated as both, these cannot be used\n",
    "                n += [annotation for annotation in SNP_anotations if annotation == \"missense_variant\"]\n",
    "                s += [annotation for annotation in SNP_anotations if annotation == \"synonymous_variant\"]\n",
    "            if n and s:\n",
    "                self.dn_ds[gene] = len(n)/len(s)\n",
    "            else:\n",
    "                self.dn_ds[gene] = \"NA\"\n",
    "            \n",
    "                \n",
    "                    \n",
    "                \n",
    "            \n",
    "        \n",
    "        \n",
    "        \n",
    "            \n",
    "\n",
    "\n",
    "gene_annotation = pd.read_csv(annotation_input,engine=\"c\",index_col=0, sep=\"\\t\")\n",
    "\n",
    "#expressed_genes = set(gene_annotation.index)\n",
    "\n",
    "\n",
    "\n",
    "impact = {\"MODIFIER\":0, \"LOW\": 0, \"MODERATE\": 0, \"HIGH\": 0 }\n",
    "locus = [33181340, 36970860]#3H locus boundaries on 3H\n",
    "\n",
    "for line in open(input_file):\n",
    "    #print(line)\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    #print(line)\n",
    "    position = int(line.rstrip().split(\"\\t\")[1])\n",
    "    if position < locus[0] or position > locus[1]:#ONly include snps in 3H locus\n",
    "        continue\n",
    "    snpeff = SnpEff(line)\n",
    "\n",
    "snpeff.clean_up_annotations()\n",
    "\n",
    "#print(snpeff)\n",
    "#print(f\"Length of Instances is {len(snpeff)}\")\n",
    "\n",
    "snpeff.get_dn_ds()#Get dn/ds ratio for each gene\n",
    "\n",
    "snpeff.write_tsv(\"/cluster/db/mecoulter/RNAseq2/3H_locus_snpeff_gene_annotation.tab\", gene_annotation)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_info = \"/cluster/db/mecoulter/RNAseq2/RNAquant_analysis_result/all_3H_genes_expressed.csv\"\n",
    "\n",
    "snp_annotation = pd.read_csv(\"/cluster/db/mecoulter/RNAseq2/3H_locus_snpeff_gene_annotation.tab\",\\\n",
    " engine=\"c\", sep=\"\\t\")\n",
    "\n",
    "expression = pd.read_csv(expression_info, engine=\"c\", index_col=1)\n",
    "\n",
    "expressed_genes = set(expression.index)\n",
    "#snp_annotation.head()\n",
    "\n",
    "snp_annotation_expressed = snp_annotation\n",
    "\n",
    "drop = False\n",
    "for i in snp_annotation.index:\n",
    "    if str(snp_annotation.at[i,\"Gene\"]) != \"nan\":\n",
    "        if snp_annotation.at[i,\"Gene\"] not in expressed_genes:\n",
    "            snp_annotation_expressed = snp_annotation_expressed.drop([i])\n",
    "            drop = True\n",
    "        else:\n",
    "            drop = False\n",
    "    else:\n",
    "        if drop:\n",
    "            snp_annotation_expressed = snp_annotation_expressed.drop([i])\n",
    "\n",
    "snp_annotation_expressed.to_csv(\"/cluster/db/mecoulter/RNAseq2/3H_locus_snpeff_gene_annotation_expressed.tab\", sep=\"\\t\")\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Position</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>ref allele</th>\n",
       "      <th>alt allele</th>\n",
       "      <th>Panzzer Annotation</th>\n",
       "      <th>dn/ds ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>BaRT2v18chr3HG123110</td>\n",
       "      <td>34748456</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>NaN</td>\n",
       "      <td>34748574</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>NaN</td>\n",
       "      <td>34748648</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>A</td>\n",
       "      <td>AG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>NaN</td>\n",
       "      <td>34749036</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>NaN</td>\n",
       "      <td>34749044</td>\n",
       "      <td>5_prime_UTR_premature_start_codon_gain_variant...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene  Position  \\\n",
       "176  BaRT2v18chr3HG123110  34748456   \n",
       "177                   NaN  34748574   \n",
       "178                   NaN  34748648   \n",
       "179                   NaN  34749036   \n",
       "180                   NaN  34749044   \n",
       "\n",
       "                                            Annotation             ref allele  \\\n",
       "176                                5_prime_UTR_variant                      G   \n",
       "177                                5_prime_UTR_variant                      G   \n",
       "178                                5_prime_UTR_variant                      A   \n",
       "179                                5_prime_UTR_variant                      C   \n",
       "180  5_prime_UTR_premature_start_codon_gain_variant...                      C   \n",
       "\n",
       "    alt allele Panzzer Annotation  dn/ds ratio  \n",
       "176          C                NaN          5.0  \n",
       "177          T                NaN          NaN  \n",
       "178         AG                NaN          NaN  \n",
       "179          G                NaN          NaN  \n",
       "180          G                NaN          NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_annotation_expressed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>contrast</th>\n",
       "      <th>adj.pval</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>up.down</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Pannzer annotation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>target</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr3HG150360</th>\n",
       "      <td>1</td>\n",
       "      <td>HEB_124_52-HEB_124_17</td>\n",
       "      <td>2.251004e-09</td>\n",
       "      <td>-10.652320</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>496316055</td>\n",
       "      <td>496319769</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr3HG152530</th>\n",
       "      <td>2</td>\n",
       "      <td>HEB_124_52-HEB_124_17</td>\n",
       "      <td>2.251004e-09</td>\n",
       "      <td>-11.909750</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>512534450</td>\n",
       "      <td>512536358</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr3HG120970</th>\n",
       "      <td>3</td>\n",
       "      <td>HEB_124_52-HEB_124_17</td>\n",
       "      <td>9.944333e-09</td>\n",
       "      <td>-5.106883</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>22786096</td>\n",
       "      <td>22788049</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr3HG121930</th>\n",
       "      <td>4</td>\n",
       "      <td>HEB_124_52-HEB_124_17</td>\n",
       "      <td>9.944333e-09</td>\n",
       "      <td>-10.847032</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>26903372</td>\n",
       "      <td>26907471</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr3HG122380</th>\n",
       "      <td>5</td>\n",
       "      <td>HEB_124_52-HEB_124_17</td>\n",
       "      <td>9.944333e-09</td>\n",
       "      <td>-9.825658</td>\n",
       "      <td>down-regulated</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>29462547</td>\n",
       "      <td>29468929</td>\n",
       "      <td>-</td>\n",
       "      <td>PHB domain-containing protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Unnamed: 0               contrast      adj.pval  \\\n",
       "target                                                                  \n",
       "BaRT2v18chr3HG150360           1  HEB_124_52-HEB_124_17  2.251004e-09   \n",
       "BaRT2v18chr3HG152530           2  HEB_124_52-HEB_124_17  2.251004e-09   \n",
       "BaRT2v18chr3HG120970           3  HEB_124_52-HEB_124_17  9.944333e-09   \n",
       "BaRT2v18chr3HG121930           4  HEB_124_52-HEB_124_17  9.944333e-09   \n",
       "BaRT2v18chr3HG122380           5  HEB_124_52-HEB_124_17  9.944333e-09   \n",
       "\n",
       "                         log2FC         up.down Chromosome      Start  \\\n",
       "target                                                                  \n",
       "BaRT2v18chr3HG150360 -10.652320  down-regulated      chr3H  496316055   \n",
       "BaRT2v18chr3HG152530 -11.909750  down-regulated      chr3H  512534450   \n",
       "BaRT2v18chr3HG120970  -5.106883  down-regulated      chr3H   22786096   \n",
       "BaRT2v18chr3HG121930 -10.847032  down-regulated      chr3H   26903372   \n",
       "BaRT2v18chr3HG122380  -9.825658  down-regulated      chr3H   29462547   \n",
       "\n",
       "                            End Strand             Pannzer annotation  \n",
       "target                                                                 \n",
       "BaRT2v18chr3HG150360  496319769      -                            NaN  \n",
       "BaRT2v18chr3HG152530  512536358      +                            NaN  \n",
       "BaRT2v18chr3HG120970   22788049      -                            NaN  \n",
       "BaRT2v18chr3HG121930   26907471      -                            NaN  \n",
       "BaRT2v18chr3HG122380   29468929      -  PHB domain-containing protein  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['BaRT2v18chr1HG000020', 'BaRT2v18chr1HG000040', 'BaRT2v18chr1HG000050',\n",
       "       'BaRT2v18chr1HG000060', 'BaRT2v18chr1HG000070'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expressed_genes[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
