{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run primers designed by Ross with BLAST to find spurious matches and for general sanity check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/cluster/db/mecoulter/Ross/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* The script we are running is this one (see below). This predicts possible products from primer blast results I made changes so that BLAST is run within the script. Lets see if it works (I include the text of the script just for reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "import sys\n",
      "import subprocess\n",
      "\n",
      "\n",
      "input_fasta = sys.argv[1]\n",
      "genome = sys.argv[2]\n",
      "outputfile = sys.argv[3]\n",
      "\n",
      "#Parameters\n",
      "maxproductsize = 4000\n",
      "max_mismatches = 3\n",
      "min_percentid = 90\n",
      "\n",
      "#header_outfmt10 = [\"name\", \"chromosome\", \"percent id\", \"match_length\", \"mismatches\", \"gap_openings\", \"Query start\",\"Query end\", \"subject_start\", \"subject_end\", \"evalue\",\"bitscore\"]\n",
      "class Blast10:#Different versions of blast produce different fields with outfmt10. So check\n",
      "\tdef __init__(self,line):\n",
      "\t\tfields = line.rstrip().split(\",\")\n",
      "\t\tself.name = fields[0]\n",
      "\t\tself.chromosome = fields[1]\n",
      "\t\tself.percent_id = float(fields[2])\n",
      "\t\tself.match_length = int(fields[3])#Excluding gaps\n",
      "\t\tself.mismatches = int(fields[4])\n",
      "\t\tself.gap_openings = int(fields[5])\n",
      "\t\tself.query_start = int(fields[6])\n",
      "\t\tself.query_end = int(fields[7])\n",
      "\t\tself.subject_start = int(fields[8])\n",
      "\t\tself.subject_end = int(fields[9])\n",
      "\t\tself.evalue = float(fields[10])\n",
      "\t\tself.bitscore = float(fields[11])\n",
      "\t\tif self.subject_start < self.subject_end:\n",
      "\t\t\tself.matchtype = \"+\"\n",
      "\t\telse:\n",
      "\t\t\tself.matchtype = \"-\"\n",
      "\t\t\t\n",
      "\n",
      "#First get lengths of primer sequences\n",
      "length_dict = {}\n",
      "for line in open(input_fasta):\n",
      "\tline = line.rstrip()\n",
      "\t#print(line)\n",
      "\tif line.startswith(\">\"):\n",
      "\t\tprimer = line.replace(\">\",\"\")\n",
      "\telse:\n",
      "\t\tlength_dict[primer] = len(line)\n",
      "\n",
      "#Get genome name\n",
      "\n",
      "genome_name = genome.split(\"/\")[-1].split(\".\")[0]\n",
      "fasta_name = input_fasta.split(\"/\")[-1].split(\".\")[0]\n",
      "print(f\"Working on genome {genome_name}\")\n",
      "\n",
      "#Run Blastn\n",
      "try:\n",
      "\tsubprocess.run(f\"blastn -task blastn-short -query {input_fasta}  -db {genome} -outfmt 10 -out {fasta_name}_{genome_name}_blast.csv\", shell=True, check=True)\n",
      "except:\n",
      "\tsubprocess.run(f\"makeblastdb -in {genome} -dbtype nucl\", shell=True, check=True)\n",
      "\tsubprocess.run(f\"blastn -task blastn-short -query {input_fasta}  -db {genome} -outfmt 10 -out {fasta_name}_{genome_name}_blast.csv\", shell=True, check=True)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "primer_matchesF = {}\n",
      "\n",
      "primer_matchesR = {}\n",
      "\n",
      "for line in open(f\"{fasta_name}_{genome_name}_blast.csv\"):\n",
      "\tblast = Blast10(line)\n",
      "\tif blast.name.endswith(\"F\"):\n",
      "\t\tprimer_matchesF.setdefault(blast.name,[]).append(blast)\n",
      "\telif blast.name.endswith(\"R\"):\n",
      "\t\tprimer_matchesR.setdefault(blast.name,[]).append(blast)\n",
      "\telse:\n",
      "\t\t\"Primer name error!\"\n",
      "\t\tsys.exit()\n",
      "\n",
      "#Go through each primer pair, filter out primers that don't have good hits and find products\n",
      "#out.write(f\"{primer_number}\\t{matchF.chromosome}\\t{productsize}\\t{start}\\t{end}\\t{matchF.percent_id}\\t{matchF.match_length}\\t{matchF.alignment_length}\\t{matchR.percent_id}\\t{matchR.match_length}\\t{matchR.alignment_length}\\n\")\n",
      "with open(outputfile, \"w\") as out:\n",
      "\tout.write(\"Primer name\\tchromosome\\tProduct size\\tStart\\tEnd\\tF percent id\\tF match length\\t F query length\\tR percent id\\tR match length\\t R query length\\n\")\n",
      "\tfor primerF, matchesF in primer_matchesF.items():\n",
      "\t\tprimerR = primerF.replace(\"F\",\"R\")\n",
      "\t\tprimer_number = primerF.replace(\"F\",\"\")\n",
      "\t\tmatchesR = primer_matchesR[primerR]\n",
      "\t\t#Filter matches\n",
      "\t\tpossible_matchesF = [matchF for matchF in matchesF if matchF.percent_id >= min_percentid and matchF.match_length >= length_dict[primerF] - max_mismatches]\n",
      "\t\tpossible_matchesR = [matchR for matchR in matchesR if matchR.percent_id >= min_percentid and matchR.match_length >= length_dict[primerF] - max_mismatches]\n",
      "\t\tfor matchF in possible_matchesF:\n",
      "\t\t\tfor matchR in possible_matchesR:\n",
      "\t\t\t\tif matchF.chromosome == matchR.chromosome and matchF.matchtype != matchR.matchtype:\n",
      "\t\t\t\t\tif matchF.matchtype == \"+\":\n",
      "\t\t\t\t\t\tproductsize = matchR.subject_end - matchF.subject_start\n",
      "\t\t\t\t\t\tstart = matchF.subject_start\n",
      "\t\t\t\t\t\tend = matchR.subject_end\n",
      "\t\t\t\t\telse:\n",
      "\t\t\t\t\t\tproductsize = matchF.subject_end - matchR.subject_start\n",
      "\t\t\t\t\t\tstart = matchR.subject_start\n",
      "\t\t\t\t\t\tend = matchF.subject_end\n",
      "\t\t\t\t\tif productsize <= maxproductsize and productsize > 40:\n",
      "\t\t\t\t\t\tout.write(f\"{primer_number}\\t{matchF.chromosome}\\t{productsize}\\t{start}\\t{end}\\t{matchF.percent_id}\\t{matchF.match_length}\\t{(matchF.query_end - matchF.query_start) + 1}\\t{matchR.percent_id}\\t{matchR.match_length}\\t{(matchR.query_end - matchR.query_start) + 1}\\n\")\n",
      "\n",
      "print(\"Script finished!\")\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat predict_products.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* I had to make several changes to the above as blast in jupyter is different to my blast on cluster. So it will not be backwards compatible (I will have to redo the previous runs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* We will run with primers Ross designed for the third intron"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on genome 180903_Barke_Unfiltered_chloro_clean_pseudomolecules_v1\n",
      "Script finished!\n"
     ]
    }
   ],
   "source": [
    "!python predict_products.py third_intron_primer_sequences.fasta /cluster/db/mecoulter/genomes/genomes/180903_Barke_Unfiltered_chloro_clean_pseudomolecules_v1.fasta third_intron_primers_Barke_blast_products.tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_csv(\"third_intron_primers_Barke_blast_products.tab\", delimiter=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Product size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>F percent id</th>\n",
       "      <th>F match length</th>\n",
       "      <th>F query length</th>\n",
       "      <th>R percent id</th>\n",
       "      <th>R match length</th>\n",
       "      <th>R query length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primer name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>intron_3_1</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>963</td>\n",
       "      <td>36887083</td>\n",
       "      <td>36888046</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_2</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>834</td>\n",
       "      <td>36887223</td>\n",
       "      <td>36888057</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_3</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>896</td>\n",
       "      <td>36887332</td>\n",
       "      <td>36888228</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_4</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>956</td>\n",
       "      <td>36887068</td>\n",
       "      <td>36888024</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_5</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>834</td>\n",
       "      <td>36887222</td>\n",
       "      <td>36888056</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_6</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>973</td>\n",
       "      <td>36887081</td>\n",
       "      <td>36888054</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_7</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>906</td>\n",
       "      <td>36887149</td>\n",
       "      <td>36888055</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_8</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>889</td>\n",
       "      <td>36887331</td>\n",
       "      <td>36888220</td>\n",
       "      <td>100.0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_9</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>823</td>\n",
       "      <td>36887221</td>\n",
       "      <td>36888044</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_10</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>892</td>\n",
       "      <td>36887147</td>\n",
       "      <td>36888039</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chromosome  Product size     Start       End  F percent id  \\\n",
       "Primer name                                                              \n",
       "intron_3_1       chr3H           963  36887083  36888046         100.0   \n",
       "intron_3_2       chr3H           834  36887223  36888057         100.0   \n",
       "intron_3_3       chr3H           896  36887332  36888228         100.0   \n",
       "intron_3_4       chr3H           956  36887068  36888024         100.0   \n",
       "intron_3_5       chr3H           834  36887222  36888056         100.0   \n",
       "intron_3_6       chr3H           973  36887081  36888054         100.0   \n",
       "intron_3_7       chr3H           906  36887149  36888055         100.0   \n",
       "intron_3_8       chr3H           889  36887331  36888220         100.0   \n",
       "intron_3_9       chr3H           823  36887221  36888044         100.0   \n",
       "intron_3_10      chr3H           892  36887147  36888039         100.0   \n",
       "\n",
       "             F match length   F query length  R percent id  R match length  \\\n",
       "Primer name                                                                  \n",
       "intron_3_1               23               23         100.0              23   \n",
       "intron_3_2               23               23         100.0              23   \n",
       "intron_3_3               23               23         100.0              23   \n",
       "intron_3_4               23               23         100.0              23   \n",
       "intron_3_5               23               23         100.0              23   \n",
       "intron_3_6               23               23         100.0              23   \n",
       "intron_3_7               23               23         100.0              23   \n",
       "intron_3_8               24               24         100.0              23   \n",
       "intron_3_9               23               23         100.0              23   \n",
       "intron_3_10              23               23         100.0              23   \n",
       "\n",
       "              R query length  \n",
       "Primer name                   \n",
       "intron_3_1                23  \n",
       "intron_3_2                23  \n",
       "intron_3_3                23  \n",
       "intron_3_4                23  \n",
       "intron_3_5                23  \n",
       "intron_3_6                23  \n",
       "intron_3_7                23  \n",
       "intron_3_8                23  \n",
       "intron_3_9                23  \n",
       "intron_3_10               23  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Run on Morex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on genome Morex_pseudomolecules_v2\n",
      "BLAST Database error: Error: Not a valid version 4 database.\n",
      "Blastn database not found. Making one...\n",
      "\n",
      "\n",
      "Building a new DB, current time: 07/06/2021 12:38:05\n",
      "New DB name:   /cluster/db/mecoulter/genomes/genomes/Morex_pseudomolecules_v2.fasta\n",
      "New DB title:  /cluster/db/mecoulter/genomes/genomes/Morex_pseudomolecules_v2.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /cluster/db/mecoulter/genomes/genomes/Morex_pseudomolecules_v2.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 8 sequences in 95.0353 seconds.\n",
      "Script finished!\n"
     ]
    }
   ],
   "source": [
    "!python predict_products.py third_intron_primer_sequences.fasta /cluster/db/mecoulter/genomes/genomes/Morex_pseudomolecules_v2.fasta third_intron_primers_Morex_blast_products.tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_csv(\"third_intron_primers_Morex_blast_products.tab\", delimiter=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Product size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>F percent id</th>\n",
       "      <th>F match length</th>\n",
       "      <th>F query length</th>\n",
       "      <th>R percent id</th>\n",
       "      <th>R match length</th>\n",
       "      <th>R query length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primer name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chromosome, Product size, Start, End, F percent id, F match length,  F query length, R percent id, R match length,  R query length]\n",
       "Index: []"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* No hits as expected (Morex does not have the NLR). Now try with Golden Promise (This should work)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on genome Golden_Promise_pseudomolecule_v1\n",
      "BLAST Database error: Error: Not a valid version 4 database.\n",
      "Blastn database not found. Making one...\n",
      "\n",
      "\n",
      "Building a new DB, current time: 07/06/2021 12:43:19\n",
      "New DB name:   /cluster/db/mecoulter/genomes/genomes/Golden_Promise_pseudomolecule_v1.fasta\n",
      "New DB title:  /cluster/db/mecoulter/genomes/genomes/Golden_Promise_pseudomolecule_v1.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /cluster/db/mecoulter/genomes/genomes/Golden_Promise_pseudomolecule_v1.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 8 sequences in 90.1037 seconds.\n",
      "Script finished!\n"
     ]
    }
   ],
   "source": [
    "!python predict_products.py third_intron_primer_sequences.fasta /cluster/db/mecoulter/genomes/genomes/Golden_Promise_pseudomolecule_v1.fasta third_intron_primers_Golden_Promise_blast_products.tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_csv(\"third_intron_primers_Golden_Promise_blast_products.tab\", delimiter=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Product size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>F percent id</th>\n",
       "      <th>F match length</th>\n",
       "      <th>F query length</th>\n",
       "      <th>R percent id</th>\n",
       "      <th>R match length</th>\n",
       "      <th>R query length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primer name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>intron_3_1</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>963</td>\n",
       "      <td>30345568</td>\n",
       "      <td>30346531</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_2</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>834</td>\n",
       "      <td>30345708</td>\n",
       "      <td>30346542</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_3</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>896</td>\n",
       "      <td>30345817</td>\n",
       "      <td>30346713</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_4</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>956</td>\n",
       "      <td>30345553</td>\n",
       "      <td>30346509</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_5</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>834</td>\n",
       "      <td>30345707</td>\n",
       "      <td>30346541</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_6</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>973</td>\n",
       "      <td>30345566</td>\n",
       "      <td>30346539</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_7</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>906</td>\n",
       "      <td>30345634</td>\n",
       "      <td>30346540</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_8</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>889</td>\n",
       "      <td>30345816</td>\n",
       "      <td>30346705</td>\n",
       "      <td>100.0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_9</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>823</td>\n",
       "      <td>30345706</td>\n",
       "      <td>30346529</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_3_10</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>892</td>\n",
       "      <td>30345632</td>\n",
       "      <td>30346524</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chromosome  Product size     Start       End  F percent id  \\\n",
       "Primer name                                                              \n",
       "intron_3_1       chr3H           963  30345568  30346531         100.0   \n",
       "intron_3_2       chr3H           834  30345708  30346542         100.0   \n",
       "intron_3_3       chr3H           896  30345817  30346713         100.0   \n",
       "intron_3_4       chr3H           956  30345553  30346509         100.0   \n",
       "intron_3_5       chr3H           834  30345707  30346541         100.0   \n",
       "intron_3_6       chr3H           973  30345566  30346539         100.0   \n",
       "intron_3_7       chr3H           906  30345634  30346540         100.0   \n",
       "intron_3_8       chr3H           889  30345816  30346705         100.0   \n",
       "intron_3_9       chr3H           823  30345706  30346529         100.0   \n",
       "intron_3_10      chr3H           892  30345632  30346524         100.0   \n",
       "\n",
       "             F match length   F query length  R percent id  R match length  \\\n",
       "Primer name                                                                  \n",
       "intron_3_1               23               23         100.0              23   \n",
       "intron_3_2               23               23         100.0              23   \n",
       "intron_3_3               23               23         100.0              23   \n",
       "intron_3_4               23               23         100.0              23   \n",
       "intron_3_5               23               23         100.0              23   \n",
       "intron_3_6               23               23         100.0              23   \n",
       "intron_3_7               23               23         100.0              23   \n",
       "intron_3_8               24               24         100.0              23   \n",
       "intron_3_9               23               23         100.0              23   \n",
       "intron_3_10              23               23         100.0              23   \n",
       "\n",
       "              R query length  \n",
       "Primer name                   \n",
       "intron_3_1                23  \n",
       "intron_3_2                23  \n",
       "intron_3_3                23  \n",
       "intron_3_4                23  \n",
       "intron_3_5                23  \n",
       "intron_3_6                23  \n",
       "intron_3_7                23  \n",
       "intron_3_8                23  \n",
       "intron_3_9                23  \n",
       "intron_3_10               23  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Try with B1K (we expect a negative result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "B1K-04-12_pseudomolecules_v1.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on genome B1K-04-12_pseudomolecules_v1\n",
      "BLAST Database error: Error: Not a valid version 4 database.\n",
      "Blastn database not found. Making one...\n",
      "\n",
      "\n",
      "Building a new DB, current time: 07/06/2021 12:51:59\n",
      "New DB name:   /cluster/db/mecoulter/genomes/genomes/B1K-04-12_pseudomolecules_v1.fasta\n",
      "New DB title:  /cluster/db/mecoulter/genomes/genomes/B1K-04-12_pseudomolecules_v1.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /cluster/db/mecoulter/genomes/genomes/B1K-04-12_pseudomolecules_v1.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 8 sequences in 91.8752 seconds.\n",
      "Script finished!\n"
     ]
    }
   ],
   "source": [
    "!python predict_products.py third_intron_primer_sequences.fasta /cluster/db/mecoulter/genomes/genomes/B1K-04-12_pseudomolecules_v1.fasta third_intron_primers_B1K_blast_products.tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pd.read_csv(\"third_intron_primers_B1K_blast_products.tab\", delimiter=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Product size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>F percent id</th>\n",
       "      <th>F match length</th>\n",
       "      <th>F query length</th>\n",
       "      <th>R percent id</th>\n",
       "      <th>R match length</th>\n",
       "      <th>R query length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primer name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chromosome, Product size, Start, End, F percent id, F match length,  F query length, R percent id, R match length,  R query length]\n",
       "Index: []"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Conclusion: in silico predictions are as expected, intron 3 primers all good!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
