{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 07/07/21: find lines similar to both 124_52 and 124_17 in mapping dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Steps: \n",
    "* Convert 50K genotypes of 124_52 and 124_17 into 9K genotypes\n",
    "\n",
    "* Find which have closest match in terms of genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Input files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "QTL_input = \"/cluster/db/mecoulter/misc/JH07_merged_HID144-Barke_Silva138_ASV_enriched_Mapping_0621.csv\"\n",
    "genotype_input = \"/cluster/db/mecoulter/genomes/markers/introgression_viewer_test.txt\"#50k genotypic data\n",
    "oldmarkers = \"/cluster/db/mecoulter/genomes/markers/Current_Map_Old_Names_060318.txt\"\n",
    "newmarkers = \"/cluster/db/mecoulter/genomes/markers/Current_Map_New_Names_060318.txt\"\n",
    "further_genotype_info_input = \"/cluster/db/mecoulter/genomes/markers/mapping_genotype_information.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [],
   "source": [
    "QTL_table = pd.read_csv(QTL_input,engine=\"c\")\n",
    "genotypes = pd.read_csv(genotype_input,engine=\"c\",delimiter=\"\\t\")\n",
    "old = pd.read_csv(oldmarkers,engine=\"c\",delimiter=\"\\t\",header=None)\n",
    "new = pd.read_csv(newmarkers,engine=\"c\",delimiter=\"\\t\",header=None)\n",
    "further_genotype_info = pd.read_csv(further_genotype_info_input,engine=\"c\",delimiter=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JHI-Hv50k-2016-7</td>\n",
       "      <td>1H</td>\n",
       "      <td>42931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JHI-Hv50k-2016-11</td>\n",
       "      <td>1H</td>\n",
       "      <td>43264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JHI-Hv50k-2016-24</td>\n",
       "      <td>1H</td>\n",
       "      <td>44868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JHI-Hv50k-2016-64</td>\n",
       "      <td>1H</td>\n",
       "      <td>48518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JHI-Hv50k-2016-66</td>\n",
       "      <td>1H</td>\n",
       "      <td>48668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44035</th>\n",
       "      <td>SCRI_RS_9180</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44036</th>\n",
       "      <td>SCRI_RS_92500</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44037</th>\n",
       "      <td>SCRI_RS_92998</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44038</th>\n",
       "      <td>SCRI_RS_994</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44039</th>\n",
       "      <td>SCRI_RS_9980</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44040 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       0   1      2\n",
       "0       JHI-Hv50k-2016-7  1H  42931\n",
       "1      JHI-Hv50k-2016-11  1H  43264\n",
       "2      JHI-Hv50k-2016-24  1H  44868\n",
       "3      JHI-Hv50k-2016-64  1H  48518\n",
       "4      JHI-Hv50k-2016-66  1H  48668\n",
       "...                  ...  ..    ...\n",
       "44035       SCRI_RS_9180   U      0\n",
       "44036      SCRI_RS_92500   U      0\n",
       "44037      SCRI_RS_92998   U      0\n",
       "44038        SCRI_RS_994   U      0\n",
       "44039       SCRI_RS_9980   U      0\n",
       "\n",
       "[44040 rows x 3 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Set up conversion from new to old markers as lookup table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "conversion = pd.DataFrame({\"old\": old[0], \"new\": new[0],})\n",
    "conversion = conversion.set_index(\"new\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Streptomyces3',\n",
       " 'Streptomyces4',\n",
       " 'BOPA1_2511_533',\n",
       " 'BOPA1_8670_388',\n",
       " 'BOPA2_12_10420',\n",
       " 'BOPA2_12_30653',\n",
       " 'BOPA2_12_30969',\n",
       " 'SCRI_RS_124234',\n",
       " 'SCRI_RS_138128',\n",
       " 'SCRI_RS_142714',\n",
       " 'SCRI_RS_149838',\n",
       " 'SCRI_RS_180989',\n",
       " 'SCRI_RS_204276',\n",
       " 'SCRI_RS_227491',\n",
       " 'SCRI_RS_120053',\n",
       " 'SCRI_RS_120059']"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(QTL_table.columns)[34:50]#Index of first marker is 36"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Now go through the genotype file, and convert markers to old names\n",
    "\n",
    "* Then filter markers based on whether they are present in the 9k set.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_markers = pd.Series([conversion.at[marker,\"old\"] for marker in genotypes[\"Marker\"]]) #Old name conversion\n",
    "genotypes[\"old\"] = old_markers\n",
    "mapping_markers = pd.DataFrame({\"old\":QTL_table.columns[36:]})#9K markers\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Now merge the two dataframes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_9k_subset = pd.merge(mapping_markers,genotypes, on=\"old\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>old</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOPA1_2511_533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BOPA1_8670_388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BOPA2_12_10420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BOPA2_12_30653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BOPA2_12_30969</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              old\n",
       "0  BOPA1_2511_533\n",
       "1  BOPA1_8670_388\n",
       "2  BOPA2_12_10420\n",
       "3  BOPA2_12_30653\n",
       "4  BOPA2_12_30969"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>old</th>\n",
       "      <th>Marker</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Barke position</th>\n",
       "      <th>Barke</th>\n",
       "      <th>Int_52</th>\n",
       "      <th>Int_33</th>\n",
       "      <th>Int_42</th>\n",
       "      <th>Int_17</th>\n",
       "      <th>Int_19</th>\n",
       "      <th>Int_56</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOPA2_12_10420</td>\n",
       "      <td>12_10420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BOPA2_12_30653</td>\n",
       "      <td>12_30653</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BOPA2_12_30969</td>\n",
       "      <td>12_30969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCRI_RS_124234</td>\n",
       "      <td>SCRI_RS_124234</td>\n",
       "      <td>chr1H</td>\n",
       "      <td>2913763.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCRI_RS_142714</td>\n",
       "      <td>SCRI_RS_142714</td>\n",
       "      <td>chr1H</td>\n",
       "      <td>2308450.0</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              old          Marker Chromosome  Barke position Barke Int_52  \\\n",
       "0  BOPA2_12_10420        12_10420        NaN             NaN     C      C   \n",
       "1  BOPA2_12_30653        12_30653        NaN             NaN     A      A   \n",
       "2  BOPA2_12_30969        12_30969        NaN             NaN     A      A   \n",
       "3  SCRI_RS_124234  SCRI_RS_124234      chr1H       2913763.0     T      T   \n",
       "4  SCRI_RS_142714  SCRI_RS_142714      chr1H       2308450.0     A      A   \n",
       "\n",
       "  Int_33 Int_42 Int_17 Int_19 Int_56  \n",
       "0      C      C      C      C      C  \n",
       "1      A      A      A      A      A  \n",
       "2      A      A      A      A      A  \n",
       "3      T      T      T      T      T  \n",
       "4      A      A      A      A      A  "
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_9k_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env.OPS_18</th>\n",
       "      <th>Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium</th>\n",
       "      <th>Pedobacter1</th>\n",
       "      <th>Paenibacillus1</th>\n",
       "      <th>Niastella</th>\n",
       "      <th>Paenibacillus2</th>\n",
       "      <th>Stenotrophomonas</th>\n",
       "      <th>Paenibacillus3</th>\n",
       "      <th>Lysobacter1</th>\n",
       "      <th>Asticcacaulis</th>\n",
       "      <th>...</th>\n",
       "      <th>SCRI_RS_193505</th>\n",
       "      <th>BOPA1_1493_839</th>\n",
       "      <th>SCRI_RS_123211</th>\n",
       "      <th>SCRI_RS_167617</th>\n",
       "      <th>SCRI_RS_168479</th>\n",
       "      <th>SCRI_RS_169268</th>\n",
       "      <th>SCRI_RS_214451</th>\n",
       "      <th>BOPA1_1847_1745</th>\n",
       "      <th>BOPA2_12_30826</th>\n",
       "      <th>BOPA1_2457_2346</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7H</td>\n",
       "      <td>7H</td>\n",
       "      <td>7H</td>\n",
       "      <td>7H</td>\n",
       "      <td>7H</td>\n",
       "      <td>7H</td>\n",
       "      <td>7H</td>\n",
       "      <td>7H</td>\n",
       "      <td>7H</td>\n",
       "      <td>7H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>140.6</td>\n",
       "      <td>140.7</td>\n",
       "      <td>140.7</td>\n",
       "      <td>140.7</td>\n",
       "      <td>140.7</td>\n",
       "      <td>140.7</td>\n",
       "      <td>140.7</td>\n",
       "      <td>140.9</td>\n",
       "      <td>140.9</td>\n",
       "      <td>141.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>127.5</td>\n",
       "      <td>95.5</td>\n",
       "      <td>22.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>13.5</td>\n",
       "      <td>30.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>75.0</td>\n",
       "      <td>104.5</td>\n",
       "      <td>35.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>41.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20.5</td>\n",
       "      <td>83.5</td>\n",
       "      <td>39.5</td>\n",
       "      <td>39.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.5</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 5740 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   env.OPS_18  Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium  \\\n",
       "0         NaN                                                NaN    \n",
       "1         NaN                                                NaN    \n",
       "2         0.5                                                2.0    \n",
       "3         8.5                                                0.5    \n",
       "4         0.5                                                1.0    \n",
       "\n",
       "   Pedobacter1  Paenibacillus1  Niastella  Paenibacillus2  Stenotrophomonas  \\\n",
       "0          NaN             NaN        NaN             NaN               NaN   \n",
       "1          NaN             NaN        NaN             NaN               NaN   \n",
       "2        127.5            95.5       22.0            52.0               3.5   \n",
       "3         75.0           104.5       35.0            62.0               4.0   \n",
       "4         20.5            83.5       39.5            39.5               0.0   \n",
       "\n",
       "   Paenibacillus3  Lysobacter1  Asticcacaulis  ...  SCRI_RS_193505  \\\n",
       "0             NaN          NaN            NaN  ...              7H   \n",
       "1             NaN          NaN            NaN  ...           140.6   \n",
       "2            13.5         30.5            0.5  ...               A   \n",
       "3            18.0         41.5            0.5  ...               A   \n",
       "4            18.5         72.0            0.0  ...               A   \n",
       "\n",
       "   BOPA1_1493_839  SCRI_RS_123211  SCRI_RS_167617  SCRI_RS_168479  \\\n",
       "0              7H              7H              7H              7H   \n",
       "1           140.7           140.7           140.7           140.7   \n",
       "2               -               A               -               A   \n",
       "3               -               A               -               A   \n",
       "4               -               A               -               A   \n",
       "\n",
       "   SCRI_RS_169268  SCRI_RS_214451  BOPA1_1847_1745  BOPA2_12_30826  \\\n",
       "0              7H              7H               7H              7H   \n",
       "1           140.7           140.7            140.9           140.9   \n",
       "2               A               -                A               -   \n",
       "3               A               -                A               -   \n",
       "4               A               -                A               -   \n",
       "\n",
       "   BOPA1_2457_2346  \n",
       "0               7H  \n",
       "1            141.1  \n",
       "2                -  \n",
       "3                -  \n",
       "4                -  \n",
       "\n",
       "[5 rows x 5740 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "QTL_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Produce a dataframe of similarity. Start with Int_17 as a test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "QTL_T = QTL_table.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-180-a527adf24f0b>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  QTL[\"old\"] = QTL.index\n"
     ]
    }
   ],
   "source": [
    "QTL = QTL_T\n",
    "QTL = QTL[36:]\n",
    "QTL.head()\n",
    "#There are markers in the QTL table which are not in the genotype_9k_subset\n",
    "QTL[\"old\"] = QTL.index\n",
    "QTL2 = pd.merge(genotype_9k_subset, QTL, on=\"old\")\n",
    "assert list(QTL2[\"old\"]) == list(genotype_9k_subset[\"old\"])#Check that these are both the same\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>old</th>\n",
       "      <th>Marker</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Barke position</th>\n",
       "      <th>Barke</th>\n",
       "      <th>Int_52</th>\n",
       "      <th>Int_33</th>\n",
       "      <th>Int_42</th>\n",
       "      <th>Int_17</th>\n",
       "      <th>Int_19</th>\n",
       "      <th>...</th>\n",
       "      <th>41</th>\n",
       "      <th>42</th>\n",
       "      <th>43</th>\n",
       "      <th>44</th>\n",
       "      <th>45</th>\n",
       "      <th>46</th>\n",
       "      <th>47</th>\n",
       "      <th>48</th>\n",
       "      <th>49</th>\n",
       "      <th>50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOPA2_12_10420</td>\n",
       "      <td>12_10420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BOPA2_12_30653</td>\n",
       "      <td>12_30653</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BOPA2_12_30969</td>\n",
       "      <td>12_30969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCRI_RS_124234</td>\n",
       "      <td>SCRI_RS_124234</td>\n",
       "      <td>chr1H</td>\n",
       "      <td>2913763.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCRI_RS_142714</td>\n",
       "      <td>SCRI_RS_142714</td>\n",
       "      <td>chr1H</td>\n",
       "      <td>2308450.0</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              old          Marker Chromosome  Barke position Barke Int_52  \\\n",
       "0  BOPA2_12_10420        12_10420        NaN             NaN     C      C   \n",
       "1  BOPA2_12_30653        12_30653        NaN             NaN     A      A   \n",
       "2  BOPA2_12_30969        12_30969        NaN             NaN     A      A   \n",
       "3  SCRI_RS_124234  SCRI_RS_124234      chr1H       2913763.0     T      T   \n",
       "4  SCRI_RS_142714  SCRI_RS_142714      chr1H       2308450.0     A      A   \n",
       "\n",
       "  Int_33 Int_42 Int_17 Int_19  ... 41 42 43 44 45 46 47 48 49 50  \n",
       "0      C      C      C      C  ...  A  A  A  A  A  A  A  B  B  H  \n",
       "1      A      A      A      A  ...  A  A  A  A  A  A  A  B  B  H  \n",
       "2      A      A      A      A  ...  A  A  A  A  A  A  A  B  B  H  \n",
       "3      T      T      T      T  ...  -  -  -  -  -  -  -  -  -  -  \n",
       "4      A      A      A      A  ...  -  -  -  -  -  -  -  -  -  -  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "QTL2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Problem - data we have for introgression lines have alleles as A, T etc. We need A, B (ref alternative). So get this information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Convert lines to \"A\" and \"B\" genotypes. A is reference (Barke), \"B\" is wild allele."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>old</th>\n",
       "      <th>Marker</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Barke position</th>\n",
       "      <th>Barke</th>\n",
       "      <th>Int_52</th>\n",
       "      <th>Int_33</th>\n",
       "      <th>Int_42</th>\n",
       "      <th>Int_17</th>\n",
       "      <th>Int_19</th>\n",
       "      <th>Int_56</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BOPA2_12_10420</td>\n",
       "      <td>12_10420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BOPA2_12_30653</td>\n",
       "      <td>12_30653</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BOPA2_12_30969</td>\n",
       "      <td>12_30969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCRI_RS_124234</td>\n",
       "      <td>SCRI_RS_124234</td>\n",
       "      <td>chr1H</td>\n",
       "      <td>2913763.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCRI_RS_142714</td>\n",
       "      <td>SCRI_RS_142714</td>\n",
       "      <td>chr1H</td>\n",
       "      <td>2308450.0</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              old          Marker Chromosome  Barke position Barke Int_52  \\\n",
       "0  BOPA2_12_10420        12_10420        NaN             NaN     C      C   \n",
       "1  BOPA2_12_30653        12_30653        NaN             NaN     A      A   \n",
       "2  BOPA2_12_30969        12_30969        NaN             NaN     A      A   \n",
       "3  SCRI_RS_124234  SCRI_RS_124234      chr1H       2913763.0     T      T   \n",
       "4  SCRI_RS_142714  SCRI_RS_142714      chr1H       2308450.0     A      A   \n",
       "\n",
       "  Int_33 Int_42 Int_17 Int_19 Int_56  \n",
       "0      C      C      C      C      C  \n",
       "1      A      A      A      A      A  \n",
       "2      A      A      A      A      A  \n",
       "3      T      T      T      T      T  \n",
       "4      A      A      A      A      A  "
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_9k_subset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "name = \"Int_52\"\n",
    "line_of_interest = QTL2[name]\n",
    "Barke = QTL2[\"Barke\"]\n",
    "bases = {\"A\",\"T\",\"C\",\"G\"}\n",
    "converted = []\n",
    "for i in line_of_interest.index:\n",
    "    #print(f\"{Int_17[i]}, {Barke[i]}\")\n",
    "    if line_of_interest[i] == Barke[i]:\n",
    "        converted.append(\"A\")\n",
    "    elif line_of_interest[i] in bases:\n",
    "        converted.append(\"B\")\n",
    "    elif \"/\" in str(line_of_interest[i]) and str(Barke[i]) in str(line_of_interest[i]):#If het and ref allele in call\n",
    "        converted.append(\"H\")\n",
    "    else:\n",
    "        converted.append(\"-\")\n",
    "\n",
    "line_of_interest_c = pd.Series(converted,index=line_of_interest.index)       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "similarity_matrix = pd.DataFrame({\"A\":[1,0.5,0], \"H\": [0.5,1,0.5], \"B\": [0, 0.5, 1]}, index=[\"A\", \"H\", \"B\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* First make a similarity matrix to refer to. E.g an A allele compared to an A allele is the same (1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>H</th>\n",
       "      <th>B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>0.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     A    H    B\n",
       "A  1.0  0.5  0.0\n",
       "H  0.5  1.0  0.5\n",
       "B  0.0  0.5  1.0"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similarity_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Now go though each of the mapping lines and calculate a score out of 1 for similarity against Int_17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "similarities = {}\n",
    "possible_alleles = {\"A\",\"B\",\"H\"}\n",
    "for i in range(2,51):\n",
    "    similarity = 0\n",
    "    total = 0\n",
    "    line = i\n",
    "    for n, allele in enumerate(line_of_interest_c):\n",
    "        #print(f\"Int 17 allele is {allele}\")\n",
    "        line_allele = QTL2.at[n,line]\n",
    "        #print(f\"For marker {marker}, allele_AB is {allele_AB} and line allele is {line_allele}\")\n",
    "        #print(f\"line allele is {line_allele}\")\n",
    "        if line_allele not in possible_alleles or allele not in possible_alleles:\n",
    "            #print(f\"Error! line allele is |{line_allele}|\")\n",
    "            continue\n",
    "        similarity += similarity_matrix.at[line_allele, allele]\n",
    "        total += 1\n",
    "    try:\n",
    "        frequency = similarity/total\n",
    "    except:\n",
    "        frequency = 0\n",
    "    similarities.setdefault(name, []).append(frequency)\n",
    "\n",
    "similarities = pd.DataFrame(similarities, index=[i for i in range(2,51)])\n",
    "        \n",
    "        \n",
    "\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Sanity check - check the line with 95% id - it should be from the same family?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'HEB_15_127'"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "further_genotype_info.head()\n",
    "further_genotype_info.at[39,\"Line\"]#QTL mapping lines start at 2, Carmens info starts at 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* To complete, match up the names with the numbers...Numbers go from 2-50, while index of names is 0-48"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Int_52</th>\n",
       "      <th>Line</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>32</td>\n",
       "      <td>0.560637</td>\n",
       "      <td>HEB_15_104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>33</td>\n",
       "      <td>0.656337</td>\n",
       "      <td>HEB_15_107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>34</td>\n",
       "      <td>0.656592</td>\n",
       "      <td>HEB_15_109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>35</td>\n",
       "      <td>0.727359</td>\n",
       "      <td>HEB_15_113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>36</td>\n",
       "      <td>0.544265</td>\n",
       "      <td>HEB_15_118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>37</td>\n",
       "      <td>0.618596</td>\n",
       "      <td>HEB_15_120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>38</td>\n",
       "      <td>0.879915</td>\n",
       "      <td>HEB_15_122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>39</td>\n",
       "      <td>0.956036</td>\n",
       "      <td>HEB_15_124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>40</td>\n",
       "      <td>0.873936</td>\n",
       "      <td>HEB_15_125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>41</td>\n",
       "      <td>0.874051</td>\n",
       "      <td>HEB_15_127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>42</td>\n",
       "      <td>0.842404</td>\n",
       "      <td>HEB_15_128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index    Int_52        Line\n",
       "30     32  0.560637  HEB_15_104\n",
       "31     33  0.656337  HEB_15_107\n",
       "32     34  0.656592  HEB_15_109\n",
       "33     35  0.727359  HEB_15_113\n",
       "34     36  0.544265  HEB_15_118\n",
       "35     37  0.618596  HEB_15_120\n",
       "36     38  0.879915  HEB_15_122\n",
       "37     39  0.956036  HEB_15_124\n",
       "38     40  0.873936  HEB_15_125\n",
       "39     41  0.874051  HEB_15_127\n",
       "40     42  0.842404  HEB_15_128"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "further_genotype_info.head()\n",
    "similarities2 = similarities.reset_index()#So index is 0-48 instead of 2-50\n",
    "further_genotype_info.head()\n",
    "similarities2[\"Line\"] = further_genotype_info[\"Line\"]\n",
    "similarities2.loc[30:40]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9560361216730038"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(similarities2[\"Int_52\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Success! The line with 95% similarity is 124. The line we looked at was 124_52. This shows our method has worked well"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
