{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8370f0a5-95f3-4cf6-b353-a396b9b1bec3",
   "metadata": {},
   "source": [
    "# Work out underlying alleles of DE genes\n",
    "\n",
    "We have 50 DE genes of interest. These will either be in regions where the two lines are the same genetically, or possibly different. So lets find out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b2f993a-cde4-4ca0-8566-de2d9d391077",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of DE genes that are definately trans regulated: 6\n",
      "\n",
      "Number of DE genes on 3H: 31\n",
      "\n",
      "Number of DE genes definatetively trans regulated and on 3H: 3\n"
     ]
    }
   ],
   "source": [
    "#import pandas as pd #Ok pandas not working for the moment\n",
    "\n",
    "import pdb \n",
    "\n",
    "\n",
    "class Introgression_bed:\n",
    "    \n",
    "    def __init__(self,line):\n",
    "        \n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        self.line = line\n",
    "        self.chromosome = fields[0]\n",
    "        self.start = int(fields[1])\n",
    "        self.end = int(fields[2])\n",
    "        self.name = fields[3]\n",
    "        self.allele = fields[4]\n",
    "\n",
    "        \n",
    "\n",
    "class Bedlist:\n",
    "    \n",
    "    def __init__(self):\n",
    "        \n",
    "        self.bedlist = []\n",
    "    \n",
    "    def __len__(self):\n",
    "        \n",
    "        return len(self.bedlist)\n",
    "    \n",
    "    def __getitem__(self,index):\n",
    "        \n",
    "        return self.bedlist[index]\n",
    "        \n",
    "    def append(self,item):\n",
    "        \n",
    "        self.bedlist.append(item) \n",
    "        \n",
    "    def get_allele(self, chromosome, start, end):\n",
    "\n",
    "        \"\"\"Pulls out allele at a particular locus, denoted \n",
    "        by start and end\"\"\"\n",
    "\n",
    "        allele = \"\"\n",
    "\n",
    "        for bed in self.bedlist:\n",
    "            if chromosome ==  bed.chromosome:\n",
    "                if start > bed.start and end < bed.end:\n",
    "                    allele = bed.allele\n",
    "\n",
    "        if allele:\n",
    "            return allele\n",
    "\n",
    "        else:\n",
    "            print(\"Error! alelle not found!\")\n",
    "            return allele\n",
    "\n",
    "\n",
    "\n",
    "Int17_file = \"/cluster/db/mecoulter/RNAseq2/introgression_viewer/Int_17-Barke.bed\" #Chromosome, start, end, name, allele\n",
    "\n",
    "Int52_file = \"/cluster/db/mecoulter/RNAseq2/introgression_viewer/Int_52-Barke.bed\"\n",
    "\n",
    "#candidate_file = \"/cluster/db/mecoulter/RNAseq2/RNAquant_analysis_result/candidate_gene_complete_info50.csv\"\n",
    "candidate_file = \"/cluster/db/mecoulter/RNAseq2/RNAquant_analysis_result/candidate_gene_complete_info34_experiment2only.csv\"\n",
    "\n",
    "output = \"/cluster/db/mecoulter/RNAseq2/RNAquant_analysis_result/candidate_gene_complete_info34experiment2only_alleles.csv\"\n",
    "\n",
    "\n",
    "Int17s = Bedlist()\n",
    "Int52s = Bedlist()\n",
    "\n",
    "\n",
    "for line in open(Int17_file):\n",
    "    Int17s.append(Introgression_bed(line))\n",
    "\n",
    "for line in open(Int52_file):\n",
    "    Int52s.append(Introgression_bed(line))\n",
    "        \n",
    "\n",
    "trans_counts = set()\n",
    "chr3H_counts = set()\n",
    "\n",
    "with open(output, \"w\") as out:\n",
    "    \n",
    "    for n, line in enumerate(open(candidate_file)):\n",
    "        #print(str(n))\n",
    "        #print(line)\n",
    "        if not n:# if first line ??\n",
    "            out.write(line.strip())\n",
    "            out.write(\",124_17 allele, 124_52 allele\\n\")#Add two new columns\n",
    "            continue\n",
    "        \n",
    "        fields = line.strip().split(\",\")\n",
    "        gene = fields[1].strip('\"')\n",
    "        chromosome = fields[6].strip('\"')\n",
    "        start, end = [int(n) for n in fields[7:9]]\n",
    "        #breakpoint()\n",
    "\n",
    "        Int17_allele = Int17s.get_allele(chromosome, start, end)#Find allele for gene based on coordinates and the bed file\n",
    "        Int52_allele = Int52s.get_allele(chromosome, start, end)\n",
    "        \n",
    "        if Int17_allele ==  Int52_allele:\n",
    "            trans_counts.add(gene)\n",
    "        if chromosome == \"chr3H\":\n",
    "            chr3H_counts.add(gene)\n",
    "            \n",
    "        \n",
    "        out.write(line.strip())\n",
    "        out.write(f\",{Int17_allele},{Int52_allele}\\n\")#Add two new columns\n",
    "\n",
    "    \n",
    "print(f\"Number of DE genes that are definately trans regulated: {len(trans_counts)}\\n\")\n",
    "print(f\"Number of DE genes on 3H: {len(chr3H_counts)}\\n\")\n",
    "print(f\"Number of DE genes definatetively trans regulated and on 3H: {len(trans_counts & chr3H_counts)}\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19e886f0-0d96-4064-8f12-fbb6ff409bb0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
