{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 14/07/21\n",
    "\n",
    "* Run blast on Carmen's/Ross's second round of PCR sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "os.chdir(\"/cluster/db/mecoulter/Ross/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./results_13-7-21_run45/SQ21-182_Ank1_19_F_A01_008.seq\n",
      "./results_13-7-21_run45/SQ21-182_Ank1_19_R_B01_007.seq\n",
      "./results_13-7-21_run45/SQ21-182_Ank1_52_F_E01_004.seq\n",
      "./results_13-7-21_run45/SQ21-182_Ank1_52_R_F01_003.seq\n",
      "./results_13-7-21_run45/SQ21-182_Ank1_56_F_C01_006.seq\n",
      "./results_13-7-21_run45/SQ21-182_Ank1_56_R_D01_005.seq\n",
      "./results_13-7-21_run45/SQ21-182_LRR1_52_F_G01_002.seq\n",
      "./results_13-7-21_run45/SQ21-182_LRR1_52_R_H01_001.seq\n"
     ]
    }
   ],
   "source": [
    "!ls ./results_13-7-21_run45/*.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SQ21-182_Ank1_52_R_F01_003.seq\n",
      "SQ21-182_LRR1_52_R_H01_001.seq\n",
      "SQ21-182_Ank1_56_F_C01_006.seq\n",
      "SQ21-182_Ank1_19_F_A01_008.seq\n",
      "SQ21-182_Ank1_56_R_D01_005.seq\n",
      "SQ21-182_Ank1_19_R_B01_007.seq\n",
      "SQ21-182_Ank1_52_F_E01_004.seq\n",
      "SQ21-182_LRR1_52_F_G01_002.seq\n"
     ]
    }
   ],
   "source": [
    "seq_files = []\n",
    "for file in os.listdir(\"./results_13-7-21_run45/\"):\n",
    "    if file.startswith(\"._\") or file.endswith(\".ab1\"):\n",
    "        continue\n",
    "    print(file)\n",
    "    seq_files.append(file)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Run blast using the barke genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running BLAST on sequence from SQ21-182_Ank1_52_R_F01_003.seq\n",
      "Running BLAST on sequence from SQ21-182_LRR1_52_R_H01_001.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_56_F_C01_006.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_19_F_A01_008.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_56_R_D01_005.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_19_R_B01_007.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_52_F_E01_004.seq\n",
      "Running BLAST on sequence from SQ21-182_LRR1_52_F_G01_002.seq\n"
     ]
    }
   ],
   "source": [
    "genome = \"/cluster/db/mecoulter/genomes/genomes/180903_Barke_Unfiltered_chloro_clean_pseudomolecules_v1.fasta\"\n",
    "genome_name = genome.split(\"/\")[-1].replace(\".fasta\", \"\")\n",
    "\n",
    "\n",
    "for file in seq_files:\n",
    "    print(f\"Running BLAST on sequence from {file}\")\n",
    "    file_name = \"_\".join(file.split(\"_\")[1:4]) #e.g Ank1_52_R\n",
    "    subprocess.run(f\"blastn -task blastn-short -query ./results_13-7-21_run45/{file}  -db {genome} -outfmt 10 -out {file_name}_blast.csv\", shell=True, check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* We need to check - is there evidence of SNPs in the sequence that are not just sequence errors? Ie is this some Barke contamination or is this an actual gene present in 19/56 and 52 (for 52 from PCR looks real)\n",
    "\n",
    "* To do this, we will run blast again, but produce normal blast output showing alignments (not outfmt 10). This should give an indication how many SNPs there are in the alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running BLAST on sequence from SQ21-182_Ank1_52_R_F01_003.seq\n",
      "Running BLAST on sequence from SQ21-182_LRR1_52_R_H01_001.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_56_F_C01_006.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_19_F_A01_008.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_56_R_D01_005.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_19_R_B01_007.seq\n",
      "Running BLAST on sequence from SQ21-182_Ank1_52_F_E01_004.seq\n",
      "Running BLAST on sequence from SQ21-182_LRR1_52_F_G01_002.seq\n"
     ]
    }
   ],
   "source": [
    "for file in seq_files:\n",
    "    print(f\"Running BLAST on sequence from {file}\")\n",
    "    file_name = \"_\".join(file.split(\"_\")[1:4]) #e.g Ank1_52_R\n",
    "    subprocess.run(f\"blastn -task blastn-short -query ./results_13-7-21_run45/{file}  -db {genome} -out {file_name}_blast.txt\", shell=True, check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Now go through each blast result, to make sure NLR is the top hit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>percent id</th>\n",
       "      <th>match_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>Query start</th>\n",
       "      <th>Query end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>96.414</td>\n",
       "      <td>753</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>740</td>\n",
       "      <td>36890109</td>\n",
       "      <td>36889357</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1275.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>83.988</td>\n",
       "      <td>331</td>\n",
       "      <td>50</td>\n",
       "      <td>3</td>\n",
       "      <td>401</td>\n",
       "      <td>729</td>\n",
       "      <td>44137909</td>\n",
       "      <td>44137580</td>\n",
       "      <td>9.460000e-53</td>\n",
       "      <td>214.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>91.156</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>216</td>\n",
       "      <td>355</td>\n",
       "      <td>5172331</td>\n",
       "      <td>5172185</td>\n",
       "      <td>3.470000e-46</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>84.706</td>\n",
       "      <td>255</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "      <td>475</td>\n",
       "      <td>728</td>\n",
       "      <td>44068396</td>\n",
       "      <td>44068142</td>\n",
       "      <td>1.370000e-45</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>91.096</td>\n",
       "      <td>146</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>217</td>\n",
       "      <td>355</td>\n",
       "      <td>531071438</td>\n",
       "      <td>531071293</td>\n",
       "      <td>1.370000e-45</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      name chromosome  percent id  match_length  mismatches  gap_openings  \\\n",
       "0  Query_1      chr3H      96.414           753           5             5   \n",
       "1  Query_1      chr3H      83.988           331          50             3   \n",
       "2  Query_1      chr3H      91.156           147           6             1   \n",
       "3  Query_1      chr3H      84.706           255          38             1   \n",
       "4  Query_1      chr3H      91.096           146           6             1   \n",
       "\n",
       "   Query start  Query end  subject_start  subject_end        evalue  bitscore  \n",
       "0           10        740       36890109     36889357  0.000000e+00    1275.0  \n",
       "1          401        729       44137909     44137580  9.460000e-53     214.0  \n",
       "2          216        355        5172331      5172185  3.470000e-46     192.0  \n",
       "3          475        728       44068396     44068142  1.370000e-45     190.0  \n",
       "4          217        355      531071438    531071293  1.370000e-45     190.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "header_outfmt10 = [\"name\", \"chromosome\", \"percent id\", \"match_length\", \"mismatches\", \"gap_openings\", \"Query start\",\"Query end\", \"subject_start\", \"subject_end\", \"evalue\",\"bitscore\"]\n",
    "blast = pd.read_csv(\"Ank1_52_R_blast.csv\", header=None)\n",
    "blast.columns = header_outfmt10\n",
    "blast.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>percent id</th>\n",
       "      <th>match_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>Query start</th>\n",
       "      <th>Query end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>96.809</td>\n",
       "      <td>752</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>739</td>\n",
       "      <td>36889397</td>\n",
       "      <td>36890148</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1310.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>84.158</td>\n",
       "      <td>303</td>\n",
       "      <td>46</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>309</td>\n",
       "      <td>44137608</td>\n",
       "      <td>44137909</td>\n",
       "      <td>9.040000e-50</td>\n",
       "      <td>204.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>91.156</td>\n",
       "      <td>147</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>355</td>\n",
       "      <td>494</td>\n",
       "      <td>5172185</td>\n",
       "      <td>5172331</td>\n",
       "      <td>3.440000e-46</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>91.096</td>\n",
       "      <td>146</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>355</td>\n",
       "      <td>493</td>\n",
       "      <td>531071293</td>\n",
       "      <td>531071438</td>\n",
       "      <td>1.360000e-45</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>85.088</td>\n",
       "      <td>228</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>235</td>\n",
       "      <td>44068169</td>\n",
       "      <td>44068396</td>\n",
       "      <td>3.310000e-43</td>\n",
       "      <td>182.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      name chromosome  percent id  match_length  mismatches  gap_openings  \\\n",
       "0  Query_1      chr3H      96.809           752           4             3   \n",
       "1  Query_1      chr3H      84.158           303          46             2   \n",
       "2  Query_1      chr3H      91.156           147           6             1   \n",
       "3  Query_1      chr3H      91.096           146           6             1   \n",
       "4  Query_1      chr3H      85.088           228          34             0   \n",
       "\n",
       "   Query start  Query end  subject_start  subject_end        evalue  bitscore  \n",
       "0            8        739       36889397     36890148  0.000000e+00    1310.0  \n",
       "1            8        309       44137608     44137909  9.040000e-50     204.0  \n",
       "2          355        494        5172185      5172331  3.440000e-46     192.0  \n",
       "3          355        493      531071293    531071438  1.360000e-45     190.0  \n",
       "4            8        235       44068169     44068396  3.310000e-43     182.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast = pd.read_csv(\"Ank1_52_F_blast.csv\", header=None)\n",
    "blast.columns = header_outfmt10\n",
    "blast.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Checking alignments in Ank1_52_F_blast.txt and Ank1_52_R_blast.txt reveals large 18bp deletion compared to Barke in both F and R sequences. Other SNPs as well. Suggests sequence is indeed from line 52 and is not contamination (but would have predicted this from PCR anyway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>percent id</th>\n",
       "      <th>match_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>Query start</th>\n",
       "      <th>Query end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>98.995</td>\n",
       "      <td>796</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>807</td>\n",
       "      <td>36885547</td>\n",
       "      <td>36886342</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1505.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>90.566</td>\n",
       "      <td>53</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>646</td>\n",
       "      <td>698</td>\n",
       "      <td>192434322</td>\n",
       "      <td>192434270</td>\n",
       "      <td>5.420000e-08</td>\n",
       "      <td>65.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>79.878</td>\n",
       "      <td>164</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>99</td>\n",
       "      <td>262</td>\n",
       "      <td>192434832</td>\n",
       "      <td>192434669</td>\n",
       "      <td>2.140000e-07</td>\n",
       "      <td>63.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>91.667</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>591</td>\n",
       "      <td>626</td>\n",
       "      <td>192434370</td>\n",
       "      <td>192434335</td>\n",
       "      <td>1.300000e-02</td>\n",
       "      <td>48.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr5H</td>\n",
       "      <td>83.792</td>\n",
       "      <td>654</td>\n",
       "      <td>103</td>\n",
       "      <td>1</td>\n",
       "      <td>44</td>\n",
       "      <td>697</td>\n",
       "      <td>584203814</td>\n",
       "      <td>584204464</td>\n",
       "      <td>1.490000e-125</td>\n",
       "      <td>456.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      name chromosome  percent id  match_length  mismatches  gap_openings  \\\n",
       "0  Query_1      chr3H      98.995           796           6             2   \n",
       "1  Query_1      chr3H      90.566            53           5             0   \n",
       "2  Query_1      chr3H      79.878           164          33             0   \n",
       "3  Query_1      chr3H      91.667            36           3             0   \n",
       "4  Query_1      chr5H      83.792           654         103             1   \n",
       "\n",
       "   Query start  Query end  subject_start  subject_end         evalue  bitscore  \n",
       "0           14        807       36885547     36886342   0.000000e+00    1505.0  \n",
       "1          646        698      192434322    192434270   5.420000e-08      65.9  \n",
       "2           99        262      192434832    192434669   2.140000e-07      63.9  \n",
       "3          591        626      192434370    192434335   1.300000e-02      48.1  \n",
       "4           44        697      584203814    584204464  1.490000e-125     456.0  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast = pd.read_csv(\"LRR1_52_F_blast.csv\", header=None)\n",
    "blast.columns = header_outfmt10\n",
    "blast.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>percent id</th>\n",
       "      <th>match_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>Query start</th>\n",
       "      <th>Query end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>96.940</td>\n",
       "      <td>719</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>734</td>\n",
       "      <td>36886379</td>\n",
       "      <td>36885661</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1271.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>90.566</td>\n",
       "      <td>53</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>163</td>\n",
       "      <td>215</td>\n",
       "      <td>192434270</td>\n",
       "      <td>192434322</td>\n",
       "      <td>5.310000e-08</td>\n",
       "      <td>65.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>91.667</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>235</td>\n",
       "      <td>270</td>\n",
       "      <td>192434335</td>\n",
       "      <td>192434370</td>\n",
       "      <td>1.200000e-02</td>\n",
       "      <td>48.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>83.333</td>\n",
       "      <td>72</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>599</td>\n",
       "      <td>670</td>\n",
       "      <td>192434669</td>\n",
       "      <td>192434740</td>\n",
       "      <td>1.200000e-02</td>\n",
       "      <td>48.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>100.000</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>33</td>\n",
       "      <td>556515769</td>\n",
       "      <td>556515789</td>\n",
       "      <td>7.700000e-01</td>\n",
       "      <td>42.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      name chromosome  percent id  match_length  mismatches  gap_openings  \\\n",
       "0  Query_1      chr3H      96.940           719          22             0   \n",
       "1  Query_1      chr3H      90.566            53           5             0   \n",
       "2  Query_1      chr3H      91.667            36           3             0   \n",
       "3  Query_1      chr3H      83.333            72          12             0   \n",
       "4  Query_1      chr3H     100.000            21           0             0   \n",
       "\n",
       "   Query start  Query end  subject_start  subject_end        evalue  bitscore  \n",
       "0           16        734       36886379     36885661  0.000000e+00    1271.0  \n",
       "1          163        215      192434270    192434322  5.310000e-08      65.9  \n",
       "2          235        270      192434335    192434370  1.200000e-02      48.1  \n",
       "3          599        670      192434669    192434740  1.200000e-02      48.1  \n",
       "4           13         33      556515769    556515789  7.700000e-01      42.1  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast = pd.read_csv(\"LRR1_52_R_blast.csv\", header=None)\n",
    "blast.columns = header_outfmt10\n",
    "blast.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Alignments (LRR1_52_R_blast.txt and LRR1_52_F_blast.txt) show sequence qute similar to Barke LRR in the NLR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>percent id</th>\n",
       "      <th>match_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>Query start</th>\n",
       "      <th>Query end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>94.595</td>\n",
       "      <td>222</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>232</td>\n",
       "      <td>36889399</td>\n",
       "      <td>36889618</td>\n",
       "      <td>5.450000e-88</td>\n",
       "      <td>331.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>84.768</td>\n",
       "      <td>151</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>161</td>\n",
       "      <td>44137610</td>\n",
       "      <td>44137760</td>\n",
       "      <td>1.530000e-23</td>\n",
       "      <td>117.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>82.955</td>\n",
       "      <td>176</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>186</td>\n",
       "      <td>44068171</td>\n",
       "      <td>44068346</td>\n",
       "      <td>9.440000e-22</td>\n",
       "      <td>111.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>86.364</td>\n",
       "      <td>44</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>135</td>\n",
       "      <td>178</td>\n",
       "      <td>44134975</td>\n",
       "      <td>44135018</td>\n",
       "      <td>7.300000e-01</td>\n",
       "      <td>42.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      name chromosome  percent id  match_length  mismatches  gap_openings  \\\n",
       "0  Query_1      chr3H      94.595           222           9             3   \n",
       "1  Query_1      chr3H      84.768           151          22             1   \n",
       "2  Query_1      chr3H      82.955           176          29             1   \n",
       "3  Query_1      chr3H      86.364            44           6             0   \n",
       "\n",
       "   Query start  Query end  subject_start  subject_end        evalue  bitscore  \n",
       "0           12        232       36889399     36889618  5.450000e-88     331.0  \n",
       "1           12        161       44137610     44137760  1.530000e-23     117.0  \n",
       "2           12        186       44068171     44068346  9.440000e-22     111.0  \n",
       "3          135        178       44134975     44135018  7.300000e-01      42.1  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "blast = pd.read_csv(\"Ank1_56_F_blast.csv\", header=None)\n",
    "blast.columns = header_outfmt10\n",
    "blast.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Very short, not sure what is wrong here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACAATANAAAGGAGACTCACAAAACATGGGGATTACTTTGGGATTACTGAAGAAGGACANCCCTCTTCCCAACATAGTTT\n",
      "ATGCATTGAGATACTGGAGCAAGCTGAAAGAAGANNCCACAACTCAGAGAAGCCTCGGTTTCTCTTGCANTAGCTGGTGA\n",
      "CCATCTATGTGGAAAGTTGTTGTACCCTCGAAAATCGAAGTAATGTTTACCTACCNAAGTGTCGAGTGTTCANANTAGTA\n",
      "GCATGGTATGGATTTGCACATNCATNGTGTATTGCACGGNGAAGGNAGNTCTCCANGTNCCANTTCCCANNGNGCTACTN\n",
      "CCCTCCTTTNCGGGGANANNGNNCGCNTTGNACTNNGANCATNGNTTTNAAAANCCCGNNATAGCCCCNGTAAGCCTAAA\n",
      "AGCCANGNANGNGNNNANGGGNGTGNGANCGAAAAGCCGCTGTAACCGGNATAGCTCNGANATNANAGAAGTTNAAGNAN\n",
      "GGGGNAANTTTCCGNGGNGNGNGATTAAAANNTCCCAGAGNGGAACTNATTNTGGNNAGAAGGGGNCCNNATTCNCAAAA\n",
      "GAAANAGGGNNTCACAAGGNNNNANCATGCGNAAAGNGGCGNAGGANNGGACNGAANNGANGGGAAGGANNNNAGGGNAN\n",
      "GGGGNANGAGGAGAAAGGCGAGGNAAAGAGNGACNNNGGAGAGCNANNNANAAANAGCANNCNAGAGGTGCCACNCCNNA\n",
      "NNNGGAANGGCGNACNAAAAAAAAAGAAAAGCNNGCNGNGGGCGNGGNNGAGGGNANAGAGNGANAGNGAGNCAANNGGN\n",
      "GNGGGAANCGGACCAGCNGAGNGNANACAANAAAAANANAAAAGGACGA\n"
     ]
    }
   ],
   "source": [
    "!cat ./results_13-7-21_run45/SQ21-182_Ank1_56_F_C01_006.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>percent id</th>\n",
       "      <th>match_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>Query start</th>\n",
       "      <th>Query end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>93.478</td>\n",
       "      <td>46</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>108</td>\n",
       "      <td>36890049</td>\n",
       "      <td>36890005</td>\n",
       "      <td>8.210000e-07</td>\n",
       "      <td>61.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>90.698</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>105</td>\n",
       "      <td>44068584</td>\n",
       "      <td>44068543</td>\n",
       "      <td>1.200000e-02</td>\n",
       "      <td>48.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>90.698</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>105</td>\n",
       "      <td>44138020</td>\n",
       "      <td>44137979</td>\n",
       "      <td>1.200000e-02</td>\n",
       "      <td>48.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr4H</td>\n",
       "      <td>100.000</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>190</td>\n",
       "      <td>209</td>\n",
       "      <td>315526524</td>\n",
       "      <td>315526543</td>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>40.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      name chromosome  percent id  match_length  mismatches  gap_openings  \\\n",
       "0  Query_1      chr3H      93.478            46           2             1   \n",
       "1  Query_1      chr3H      90.698            43           3             1   \n",
       "2  Query_1      chr3H      90.698            43           3             1   \n",
       "3  Query_1      chr4H     100.000            20           0             0   \n",
       "\n",
       "   Query start  Query end  subject_start  subject_end        evalue  bitscore  \n",
       "0           63        108       36890049     36890005  8.210000e-07      61.9  \n",
       "1           63        105       44068584     44068543  1.200000e-02      48.1  \n",
       "2           63        105       44138020     44137979  1.200000e-02      48.1  \n",
       "3          190        209      315526524    315526543  3.000000e+00      40.1  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "blast = pd.read_csv(\"Ank1_56_R_blast.csv\", header=None)\n",
    "blast.columns = header_outfmt10\n",
    "blast.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Sequence weak with lots of Ns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>percent id</th>\n",
       "      <th>match_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>Query start</th>\n",
       "      <th>Query end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>85.841</td>\n",
       "      <td>565</td>\n",
       "      <td>65</td>\n",
       "      <td>15</td>\n",
       "      <td>21</td>\n",
       "      <td>583</td>\n",
       "      <td>36889414</td>\n",
       "      <td>36889965</td>\n",
       "      <td>1.460000e-125</td>\n",
       "      <td>456.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>82.667</td>\n",
       "      <td>150</td>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "      <td>358</td>\n",
       "      <td>507</td>\n",
       "      <td>608190250</td>\n",
       "      <td>608190103</td>\n",
       "      <td>9.980000e-22</td>\n",
       "      <td>111.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>83.333</td>\n",
       "      <td>114</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>358</td>\n",
       "      <td>471</td>\n",
       "      <td>225505988</td>\n",
       "      <td>225505876</td>\n",
       "      <td>9.260000e-16</td>\n",
       "      <td>91.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>84.375</td>\n",
       "      <td>96</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>376</td>\n",
       "      <td>471</td>\n",
       "      <td>586887600</td>\n",
       "      <td>586887694</td>\n",
       "      <td>8.920000e-13</td>\n",
       "      <td>81.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>82.979</td>\n",
       "      <td>94</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>359</td>\n",
       "      <td>452</td>\n",
       "      <td>5172185</td>\n",
       "      <td>5172277</td>\n",
       "      <td>8.590000e-10</td>\n",
       "      <td>71.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      name chromosome  percent id  match_length  mismatches  gap_openings  \\\n",
       "0  Query_1      chr3H      85.841           565          65            15   \n",
       "1  Query_1      chr3H      82.667           150          24             2   \n",
       "2  Query_1      chr3H      83.333           114          18             1   \n",
       "3  Query_1      chr3H      84.375            96          14             1   \n",
       "4  Query_1      chr3H      82.979            94          15             1   \n",
       "\n",
       "   Query start  Query end  subject_start  subject_end         evalue  bitscore  \n",
       "0           21        583       36889414     36889965  1.460000e-125     456.0  \n",
       "1          358        507      608190250    608190103   9.980000e-22     111.0  \n",
       "2          358        471      225505988    225505876   9.260000e-16      91.7  \n",
       "3          376        471      586887600    586887694   8.920000e-13      81.8  \n",
       "4          359        452        5172185      5172277   8.590000e-10      71.9  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast = pd.read_csv(\"Ank1_19_F_blast.csv\", header=None)\n",
    "blast.columns = header_outfmt10\n",
    "blast.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>percent id</th>\n",
       "      <th>match_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>Query start</th>\n",
       "      <th>Query end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>86.395</td>\n",
       "      <td>147</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>53</td>\n",
       "      <td>195</td>\n",
       "      <td>36890049</td>\n",
       "      <td>36889904</td>\n",
       "      <td>9.140000e-19</td>\n",
       "      <td>101.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>95.238</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>53</td>\n",
       "      <td>94</td>\n",
       "      <td>44068584</td>\n",
       "      <td>44068543</td>\n",
       "      <td>3.230000e-09</td>\n",
       "      <td>69.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>95.238</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>53</td>\n",
       "      <td>94</td>\n",
       "      <td>44138020</td>\n",
       "      <td>44137979</td>\n",
       "      <td>3.230000e-09</td>\n",
       "      <td>69.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>100.000</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>237</td>\n",
       "      <td>258</td>\n",
       "      <td>26517470</td>\n",
       "      <td>26517491</td>\n",
       "      <td>1.800000e-01</td>\n",
       "      <td>44.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Query_1</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>100.000</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>237</td>\n",
       "      <td>258</td>\n",
       "      <td>519616398</td>\n",
       "      <td>519616419</td>\n",
       "      <td>1.800000e-01</td>\n",
       "      <td>44.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      name chromosome  percent id  match_length  mismatches  gap_openings  \\\n",
       "0  Query_1      chr3H      86.395           147          15             5   \n",
       "1  Query_1      chr3H      95.238            42           2             0   \n",
       "2  Query_1      chr3H      95.238            42           2             0   \n",
       "3  Query_1      chr3H     100.000            22           0             0   \n",
       "4  Query_1      chr3H     100.000            22           0             0   \n",
       "\n",
       "   Query start  Query end  subject_start  subject_end        evalue  bitscore  \n",
       "0           53        195       36890049     36889904  9.140000e-19     101.0  \n",
       "1           53         94       44068584     44068543  3.230000e-09      69.9  \n",
       "2           53         94       44138020     44137979  3.230000e-09      69.9  \n",
       "3          237        258       26517470     26517491  1.800000e-01      44.1  \n",
       "4          237        258      519616398    519616419  1.800000e-01      44.1  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast = pd.read_csv(\"Ank1_19_R_blast.csv\", header=None)\n",
    "blast.columns = header_outfmt10\n",
    "blast.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Checked Ank1_19_R_blast.txt and Ank1_19_F_blast.txt, lots of mismatches and shortened sequence matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NTTTTGTCATCTTTGGAAGAATCAGCGTAACTCCGAGTATGATCNCTTGAGCATGTCCATTGCTACTCTTGCCTCNGTTG\n",
      "GAAACATTATCCTCATCACCCCACTCATNGCAAGGAAAATATTGAGTGAAAAACATCCTTTCCAAGGAGAGAAAAGGTAC\n",
      "TCCCTTCGCNTCAAAATNAGTGTCTGGTTCTAGTCCNATGGTTTNAAATNGNNGGCTNCGGGAAAATGNNGCGGANCCTC\n",
      "TAAATCAGCTATAGCGGAAGCTANACGCGGCTANAGCGGCNAAATTGTNCATNANCCNNTNAACGGCANNTTGCTGAAAA\n",
      "GNCTTNGCNGGGCNATAGCCGGNCTATTTAAACTACGGTTCNNGGTCTCATTTGNACTGAAGCCGCNNNAAGGAGGGAGT\n",
      "CGCNCGNTTGGTCGTATTTCAGAGGNNNGCTTNNCTCCACNGGCNAAACAANNATCANGNGCAAAACCGGCNGCTACTAA\n",
      "TTTTTANCNNTNNNCAAANTGNGGTAAANNTNCCTCGNATNNCGAGGACNNCNNCNNTCCACATAGAGGNCACCAGGNAA\n",
      "TGCCAGAGAACCGANGCTTCTNNGAAGNNGNGGNACNNANAANAAGGCTGANNCAGNANTNTNANAGNNAGACTANGTTG\n",
      "GGGAGNNGANTNCNTAAAAGGAAGANCNGAANANCCCCAAGNANGTGANNNAAATNNANNAAAANNCNANCACCNNNNAG\n",
      "GNNAANCGGNNGGAGGCCGACCNACNNNAAAAANNNNNNNNNNNGNNNNGNGCAGACGNNCGGGNGNGAAGGGGNGAAGA\n",
      "AANNGNAANNAGANGAGGAGACGAAANAAANGGCAANNAAGGAGANAANGAGG\n"
     ]
    }
   ],
   "source": [
    "!cat ./results_13-7-21_run45/SQ21-182_Ank1_19_R_B01_007.seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* A lot of Ns!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Similar story with Ank1_56_R_blast.txt and Ank1_56_F_blast.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 15/07/21"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* No evidence of a deletion in vcf file. Check mapping of reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir(\"/cluster/db/mecoulter/RNAseq2/bam_all2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "samtools view -h -b HEB_124_52.bam chr3H_part1:36889000-36890300 > HEB_124_52_Ank1.bam\n",
    "samtools index HEB_124_52_Ank1.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Result: Deletion in intron so not picked up by RNAseq. However, reads mapping, look good, evidence of expression in 52"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now check Barke bam file... is there evidence for the pan genome annotation of the gene?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "samtools view -h -b Barke.bam chr3H_part1:36880000-36890000 > Barke_NLR.bam; samtools index Barke_NLR.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Tablet shows no evidence of the BaRT annotation of the NLR (but gaps between read pairs are small). Likely pan-genome annotation is correct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate synthetic sequence with deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/cluster/db/mecoulter/Ross/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Parse the NLR fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence = \"\"\n",
    "for line in open(\"_chr3H_36880423_36891000_sequence.fasta\"):\n",
    "    if line.startswith(\">\"):\n",
    "        continue\n",
    "    line = line.strip()\n",
    "    #print(str(len(line)))\n",
    "    sequence += line\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Delete the relevant bit of sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTATAGCTCCGCTATAG\n"
     ]
    }
   ],
   "source": [
    "deletion_start = 36889835 - 36880423 + 1\n",
    "deletion_end = deletion_start + 18\n",
    "\n",
    "deletion_sequence = sequence[deletion_start:deletion_end]\n",
    "print(deletion_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G\n",
      "T\n",
      "T\n",
      "A\n",
      "T\n",
      "A\n",
      "G\n",
      "C\n",
      "T\n",
      "C\n",
      "C\n",
      "G\n",
      "C\n",
      "T\n",
      "A\n",
      "T\n",
      "A\n",
      "G\n",
      "Length of sequence: 10578, length of deleted sequence 10560\n"
     ]
    }
   ],
   "source": [
    "deleted_sequence = \"\"\n",
    "\n",
    "for i in range(0, len(sequence)):\n",
    "    if i >= deletion_start and i < deletion_end:\n",
    "        print(sequence[i])\n",
    "        continue\n",
    "    deleted_sequence += sequence[i]\n",
    "    \n",
    "print(f\"Length of sequence: {len(sequence)}, length of deleted sequence {len(deleted_sequence)}\")   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Use sequence to design primers (NCBI primer design)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import textwrap\n",
    "\n",
    "with open(\"_chr3H_36880423_36891000_sequence_with52_deletion.fasta\",\"w\") as out:\n",
    "    out.write(\">52_NLR_deletion\\n\")\n",
    "    out.write(textwrap.fill(deleted_sequence, width=60))\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Test primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward Primer must be designed between 9391 and 9435\n"
     ]
    }
   ],
   "source": [
    "print(f\"Forward Primer must be designed between {deletion_start - 22} and {deletion_start + 22}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9413"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "9391 + 22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "import sys\n",
      "import subprocess\n",
      "\n",
      "\n",
      "input_fasta = sys.argv[1]\n",
      "genome = sys.argv[2]\n",
      "outputfile = sys.argv[3]\n",
      "\n",
      "#Parameters\n",
      "maxproductsize = 4000\n",
      "max_mismatches = 3\n",
      "min_percentid = 90\n",
      "\n",
      "#header_outfmt10 = [\"name\", \"chromosome\", \"percent id\", \"match_length\", \"mismatches\", \"gap_openings\", \"Query start\",\"Query end\", \"subject_start\", \"subject_end\", \"evalue\",\"bitscore\"]\n",
      "class Blast10:#Different versions of blast produce different fields with outfmt10. So check\n",
      "\tdef __init__(self,line):\n",
      "\t\tfields = line.rstrip().split(\",\")\n",
      "\t\tself.name = fields[0]\n",
      "\t\tself.chromosome = fields[1]\n",
      "\t\tself.percent_id = float(fields[2])\n",
      "\t\tself.match_length = int(fields[3])#Excluding gaps\n",
      "\t\tself.mismatches = int(fields[4])\n",
      "\t\tself.gap_openings = int(fields[5])\n",
      "\t\tself.query_start = int(fields[6])\n",
      "\t\tself.query_end = int(fields[7])\n",
      "\t\tself.subject_start = int(fields[8])\n",
      "\t\tself.subject_end = int(fields[9])\n",
      "\t\tself.evalue = float(fields[10])\n",
      "\t\tself.bitscore = float(fields[11])\n",
      "\t\tif self.subject_start < self.subject_end:\n",
      "\t\t\tself.matchtype = \"+\"\n",
      "\t\telse:\n",
      "\t\t\tself.matchtype = \"-\"\n",
      "\t\t\t\n",
      "\n",
      "#First get lengths of primer sequences\n",
      "length_dict = {}\n",
      "for line in open(input_fasta):\n",
      "\tline = line.rstrip()\n",
      "\t#print(line)\n",
      "\tif line.startswith(\">\"):\n",
      "\t\tprimer = line.replace(\">\",\"\")\n",
      "\telse:\n",
      "\t\tlength_dict[primer] = len(line)\n",
      "\n",
      "#Get genome name\n",
      "\n",
      "genome_name = genome.split(\"/\")[-1].split(\".\")[0]\n",
      "fasta_name = input_fasta.split(\"/\")[-1].split(\".\")[0]\n",
      "print(f\"Working on genome {genome_name}\")\n",
      "\n",
      "#Run Blastn\n",
      "try:\n",
      "\tsubprocess.run(f\"blastn -task blastn-short -query {input_fasta}  -db {genome} -outfmt 10 -out {fasta_name}_{genome_name}_blast.csv\", shell=True, check=True)\n",
      "except:\n",
      "\tprint(\"Blastn database not found. Making one...\")\n",
      "\tsubprocess.run(f\"makeblastdb -in {genome} -dbtype nucl\", shell=True, check=True)\n",
      "\tsubprocess.run(f\"blastn -task blastn-short -query {input_fasta}  -db {genome} -outfmt 10 -out {fasta_name}_{genome_name}_blast.csv\", shell=True, check=True)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "primer_matchesF = {}\n",
      "\n",
      "primer_matchesR = {}\n",
      "\n",
      "for line in open(f\"{fasta_name}_{genome_name}_blast.csv\"):\n",
      "\tblast = Blast10(line)\n",
      "\tif blast.name.endswith(\"F\"):\n",
      "\t\tprimer_matchesF.setdefault(blast.name,[]).append(blast)\n",
      "\telif blast.name.endswith(\"R\"):\n",
      "\t\tprimer_matchesR.setdefault(blast.name,[]).append(blast)\n",
      "\telse:\n",
      "\t\t\"Primer name error!\"\n",
      "\t\tsys.exit()\n",
      "\n",
      "#Go through each primer pair, filter out primers that don't have good hits and find products\n",
      "#out.write(f\"{primer_number}\\t{matchF.chromosome}\\t{productsize}\\t{start}\\t{end}\\t{matchF.percent_id}\\t{matchF.match_length}\\t{matchF.alignment_length}\\t{matchR.percent_id}\\t{matchR.match_length}\\t{matchR.alignment_length}\\n\")\n",
      "with open(outputfile, \"w\") as out:\n",
      "\tout.write(\"Primer name\\tchromosome\\tProduct size\\tStart\\tEnd\\tF percent id\\tF match length\\t F query length\\tR percent id\\tR match length\\t R query length\\n\")\n",
      "\tfor primerF, matchesF in primer_matchesF.items():\n",
      "\t\tprimerR = primerF.replace(\"F\",\"R\")\n",
      "\t\tprimer_number = primerF.replace(\"F\",\"\")\n",
      "\t\tmatchesR = primer_matchesR[primerR]\n",
      "\t\t#Filter matches\n",
      "\t\tpossible_matchesF = [matchF for matchF in matchesF if matchF.percent_id >= min_percentid and matchF.match_length >= length_dict[primerF] - max_mismatches]\n",
      "\t\tpossible_matchesR = [matchR for matchR in matchesR if matchR.percent_id >= min_percentid and matchR.match_length >= length_dict[primerF] - max_mismatches]\n",
      "\t\tfor matchF in possible_matchesF:\n",
      "\t\t\tfor matchR in possible_matchesR:\n",
      "\t\t\t\tif matchF.chromosome == matchR.chromosome and matchF.matchtype != matchR.matchtype:\n",
      "\t\t\t\t\tif matchF.matchtype == \"+\":\n",
      "\t\t\t\t\t\tproductsize = matchR.subject_end - matchF.subject_start\n",
      "\t\t\t\t\t\tstart = matchF.subject_start\n",
      "\t\t\t\t\t\tend = matchR.subject_end\n",
      "\t\t\t\t\telse:\n",
      "\t\t\t\t\t\tproductsize = matchF.subject_end - matchR.subject_start\n",
      "\t\t\t\t\t\tstart = matchR.subject_start\n",
      "\t\t\t\t\t\tend = matchF.subject_end\n",
      "\t\t\t\t\tif productsize <= maxproductsize and productsize > 40:\n",
      "\t\t\t\t\t\tout.write(f\"{primer_number}\\t{matchF.chromosome}\\t{productsize}\\t{start}\\t{end}\\t{matchF.percent_id}\\t{matchF.match_length}\\t{(matchF.query_end - matchF.query_start) + 1}\\t{matchR.percent_id}\\t{matchR.match_length}\\t{(matchR.query_end - matchR.query_start) + 1}\\n\")\n",
      "\n",
      "print(\"Script finished!\")\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat predict_products.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on genome 180903_Barke_Unfiltered_chloro_clean_pseudomolecules_v1\n",
      "Script finished!\n"
     ]
    }
   ],
   "source": [
    "!python predict_products.py 3Hlocus_deletion_primer.fasta /cluster/db/mecoulter/genomes/genomes/180903_Barke_Unfiltered_chloro_clean_pseudomolecules_v1.fasta deletion_Barke_blast_products.tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Product size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>F percent id</th>\n",
       "      <th>F match length</th>\n",
       "      <th>F query length</th>\n",
       "      <th>R percent id</th>\n",
       "      <th>R match length</th>\n",
       "      <th>R query length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primer name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>226</td>\n",
       "      <td>36889817</td>\n",
       "      <td>36890043</td>\n",
       "      <td>100.0</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>100.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>550</td>\n",
       "      <td>36889818</td>\n",
       "      <td>36890368</td>\n",
       "      <td>100.0</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>100.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chromosome  Product size     Start       End  F percent id  \\\n",
       "Primer name                                                              \n",
       "1                chr3H           226  36889817  36890043         100.0   \n",
       "7                chr3H           550  36889818  36890368         100.0   \n",
       "\n",
       "             F match length   F query length  R percent id  R match length  \\\n",
       "Primer name                                                                  \n",
       "1                        19               19         100.0              22   \n",
       "7                        18               18         100.0              22   \n",
       "\n",
       "              R query length  \n",
       "Primer name                   \n",
       "1                         22  \n",
       "7                         22  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "table = pd.read_csv(\"deletion_Barke_blast_products.tab\", delimiter=\"\\t\", index_col=0)\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished writing deletion_Morex_pseudomolecules_v2_blast_products.tab\n",
      "Finished writing deletion_Golden_Promise_pseudomolecule_v1_blast_products.tab\n",
      "Finished writing deletion_B1K-04-12_pseudomolecules_v1_blast_products.tab\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "genomes = [\"/cluster/db/mecoulter/genomes/genomes/Morex_pseudomolecules_v2.fasta\", \"/cluster/db/mecoulter/genomes/genomes/Golden_Promise_pseudomolecule_v1.fasta\", \"/cluster/db/mecoulter/genomes/genomes/B1K-04-12_pseudomolecules_v1.fasta\" ]\n",
    "\n",
    "for genome in genomes:\n",
    "    genome_name = genome.split(\"/\")[-1].split(\".\")[0]\n",
    "    subprocess.run(f\"python predict_products.py 3Hlocus_deletion_primer.fasta {genome} deletion_{genome_name}_blast_products.tab\", shell=True)\n",
    "    print(f\"Finished writing deletion_{genome_name}_blast_products.tab\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Product size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>F percent id</th>\n",
       "      <th>F match length</th>\n",
       "      <th>F query length</th>\n",
       "      <th>R percent id</th>\n",
       "      <th>R match length</th>\n",
       "      <th>R query length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primer name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chromosome, Product size, Start, End, F percent id, F match length,  F query length, R percent id, R match length,  R query length]\n",
       "Index: []"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = pd.read_csv(\"deletion_Morex_pseudomolecules_v2_blast_products.tab\", delimiter=\"\\t\", index_col=0)\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Product size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>F percent id</th>\n",
       "      <th>F match length</th>\n",
       "      <th>F query length</th>\n",
       "      <th>R percent id</th>\n",
       "      <th>R match length</th>\n",
       "      <th>R query length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primer name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>226</td>\n",
       "      <td>30348302</td>\n",
       "      <td>30348528</td>\n",
       "      <td>100.0</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>100.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr3H</td>\n",
       "      <td>550</td>\n",
       "      <td>30348303</td>\n",
       "      <td>30348853</td>\n",
       "      <td>100.0</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>100.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chromosome  Product size     Start       End  F percent id  \\\n",
       "Primer name                                                              \n",
       "1                chr3H           226  30348302  30348528         100.0   \n",
       "7                chr3H           550  30348303  30348853         100.0   \n",
       "\n",
       "             F match length   F query length  R percent id  R match length  \\\n",
       "Primer name                                                                  \n",
       "1                        19               19         100.0              22   \n",
       "7                        18               18         100.0              22   \n",
       "\n",
       "              R query length  \n",
       "Primer name                   \n",
       "1                         22  \n",
       "7                         22  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = pd.read_csv(\"deletion_Golden_Promise_pseudomolecule_v1_blast_products.tab\", delimiter=\"\\t\", index_col=0)\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Product size</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>F percent id</th>\n",
       "      <th>F match length</th>\n",
       "      <th>F query length</th>\n",
       "      <th>R percent id</th>\n",
       "      <th>R match length</th>\n",
       "      <th>R query length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primer name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chromosome, Product size, Start, End, F percent id, F match length,  F query length, R percent id, R match length,  R query length]\n",
       "Index: []"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = pd.read_csv(\"deletion_B1K-04-12_pseudomolecules_v1_blast_products.tab\", delimiter=\"\\t\", index_col=0)\n",
    "table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Primers 1 and 7 may cause false positives in Barke (but perhaps not)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
