{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare variant calls and annotations from two RNA-seq experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/cluster/db/mecoulter/CEMartinez_RNAseq-209915706/results/bcftools/3H_locus_filtered.anno.vcf'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-7cd38c0f031e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0msecond_set\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_file1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;31m#Barke - Int19/Int56\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"#\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/cluster/db/mecoulter/CEMartinez_RNAseq-209915706/results/bcftools/3H_locus_filtered.anno.vcf'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "input_file1 = \"/cluster/db/mecoulter/CEMartinez_RNAseq-209915706/results/bcftools/3H_locus_filtered.anno.vcf\"\n",
    "input_file2 = \"/cluster/db/mecoulter/RNAseq2/bcftools/3H_locus_filtered.anno.vcf\"\n",
    "\n",
    "first_set = set()\n",
    "second_set = set()\n",
    "\n",
    "for line in open(input_file1):#Barke - Int19/Int56\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    fields = line.rstrip().split(\"\\t\")\n",
    "    SNP_position = fields[1]\n",
    "    ref = fields[3]\n",
    "    alt = fields[4]\n",
    "    variant_id = f\"{SNP_position}_{ref}_{alt}\"\n",
    "    first_set.add(variant_id)\n",
    "\n",
    "print(f\"{len(first_set)} unique variants in experiment one\")\n",
    "    \n",
    "for line in open(input_file2):#Barke/Int_124_17 - Int_124_52\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    fields = line.rstrip().split(\"\\t\")\n",
    "    SNP_position = fields[1]\n",
    "    ref = fields[3]\n",
    "    alt = fields[4]\n",
    "    variant_id = f\"{SNP_position}_{ref}_{alt}\"\n",
    "    second_set.add(variant_id)\n",
    "\n",
    "print(f\"{len(second_set)} unique variants in experiment two\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* What variants are shared between two experiments?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of shared variants: 407, proportion (compared to first set): 0.2600638977635783 \n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of shared variants: {len(first_set & second_set)}, proportion (compared to first set): {len(first_set & second_set)/len(first_set)} \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of expressed genes in RNAseq dataset: 19802\n",
      "\n",
      "Number of genes with each impact category: {'MODIFIER': 145, 'LOW': 72, 'MODERATE': 72, 'HIGH': 17}\n",
      "\n",
      "Number of genes with either HIGH or MODERATE variants: 74\n",
      "\n",
      "Number of genes with HIGH or MODERATE variants with each annotation category: {'missense_variant': 72, 'frameshift_variant': 9, 'splice_donor_variant&intron_variant': 2, 'disruptive_inframe_deletion': 4, 'stop_gained': 1, 'missense_variant&splice_region_variant': 3, 'splice_acceptor_variant&intron_variant': 4, 'disruptive_inframe_insertion': 1, 'conservative_inframe_insertion': 1, 'start_lost': 1, 'frameshift_variant&stop_gained': 1}\n",
      "\n",
      "Gene BaRT2v18chr3HG122620 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122640 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG122660 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122690 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122720 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122730 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122740 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122750 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122760 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122770 annotations: missense_variant,splice_donor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG122780 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122790 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122800 annotations: missense_variant,disruptive_inframe_deletion\n",
      "Gene BaRT2v18chr3HG122810 annotations: missense_variant,stop_gained,missense_variant&splice_region_variant\n",
      "Gene BaRT2v18chr3HG122810-BaRT2v18chr3HG122820 annotations: missense_variant,missense_variant&splice_region_variant\n",
      "Gene BaRT2v18chr3HG122820 annotations: missense_variant,disruptive_inframe_deletion\n",
      "Gene BaRT2v18chr3HG122830 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122860 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122870 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG122880 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122890 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG122910 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122920 annotations: missense_variant,missense_variant&splice_region_variant,splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG122930 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122970 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122990 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123000 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123010 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123020 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123030 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123050 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123040 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123060 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123070 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123080 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123090 annotations: missense_variant,splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123100 annotations: splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123110 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123120 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123140 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123150 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123160 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123170 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123220 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123240 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123230 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123250 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123260 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123270 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123280 annotations: missense_variant,disruptive_inframe_insertion\n",
      "Gene BaRT2v18chr3HG123350 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123400 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123410 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123480 annotations: missense_variant,disruptive_inframe_deletion,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123540 annotations: missense_variant,splice_donor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123550 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123600 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123610 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123630 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123650 annotations: missense_variant,conservative_inframe_insertion\n",
      "Gene BaRT2v18chr3HG123660 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123700 annotations: splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123710 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123730-BaRT2v18chr3HG123740 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123730 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123740 annotations: missense_variant,start_lost\n",
      "Gene BaRT2v18chr3HG123750 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123760 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123770 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123780 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123790 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123800 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123810 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123830 annotations: missense_variant,disruptive_inframe_deletion,frameshift_variant&stop_gained,frameshift_variant\n",
      "\n",
      "Genes with most impacting variants:\n",
      "\n",
      "Gene BaRT2v18chr3HG122640 (Rab-GAP TBC domain-containing protein, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG122810 (Glycosyltransferase, ALG3, expressed?: True): missense_variant, stop_gained, missense_variant&splice_region_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG122870 (Ankyrin-1, expressed?: False): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG122890 (nan, expressed?: False): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123110 (nan, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123400 (nan, expressed?: False): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123480 (Folate-sensitive fragile site protein Fra10Ac1, expressed?: True): missense_variant, disruptive_inframe_deletion, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123740 (TNase-like domain-containing protein, expressed?: False): missense_variant, start_lost \n",
      "\n",
      "Gene BaRT2v18chr3HG123760 (TNase-like domain-containing protein, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123770 (Cellulose synthase-like protein H1-like, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123830 (nan, expressed?: False): missense_variant, disruptive_inframe_deletion, frameshift_variant&stop_gained, frameshift_variant \n",
      "\n"
     ]
    }
   ],
   "source": [
    "input_file = input_file1 \n",
    "\n",
    "annotation_input = \"/cluster/db/mecoulter/BaRT2v18/BaRT_2_18_annotation_genes.txt\"\n",
    "\n",
    "expression_info = \"/cluster/db/mecoulter/RNAseq2/RNAquant_analysis_result/expressed_genes_both_adj_p_values.csv\"\n",
    "\n",
    "\n",
    "gene_annotation = pd.read_csv(expression_info,engine=\"c\",index_col=0)\n",
    "\n",
    "expressed_genes = set(gene_annotation.index)\n",
    "\n",
    "print(f\"Number of expressed genes in RNAseq dataset: {len(expressed_genes)}\\n\")#Sanity check\n",
    "\n",
    "\n",
    "\n",
    "impact = {\"MODIFIER\":0, \"LOW\": 0, \"MODERATE\": 0, \"HIGH\": 0 }\n",
    "\n",
    "gene_impact = {} #Per gene impact annotation\n",
    "gene_SNP_annotation = {} #Per gene SNP annotation, HIGH or MODERATE only\n",
    "\n",
    "annotation_counts = {}\n",
    "\n",
    "for line in open(input_file):\n",
    "\tif line.startswith(\"#\"):\n",
    "\t\tcontinue\n",
    "\tfields = line.rstrip().split(\"\\t\")\n",
    "\tINFO = fields[7]\n",
    "\tSNP_position = fields[1]\n",
    "\tannotations = INFO.split(\"ANN=\")[1].split(\"=\")[0].split(\",\")#List of all possible annotations\n",
    "\tfor annotation in annotations:\n",
    "\t\ttry:\n",
    "\t\t\tallele, p_annotation, putative_impact, position, gene, feature_type, transcript, transcript_type  = annotation.split(\"|\")[:8]\n",
    "\t\texcept ValueError:\n",
    "\t\t\tprint(f\"Error at position {SNP_position}! Annotation is {annotation}\")\n",
    "\t\t\tsys.exit()\n",
    "\t\tgene_impact.setdefault(gene,set()).add(putative_impact)\n",
    "\t\tif putative_impact == \"HIGH\" or putative_impact == \"MODERATE\":\n",
    "\t\t\tgene_SNP_annotation.setdefault(gene,set()).add(p_annotation)\n",
    "\n",
    "for gene, impacts in gene_impact.items():\n",
    "\tfor putative_impact in impacts:\n",
    "\t\timpact[putative_impact] += 1\n",
    "\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "\tfor annotation in annotations:\n",
    "\t\ttry:\n",
    "\t\t\tannotation_counts[annotation] += 1\n",
    "\t\texcept KeyError:\n",
    "\t\t\tannotation_counts[annotation] = 1\n",
    "\n",
    "print(f\"Number of genes with each impact category: {impact}\\n\")#One gene can have multiple categories\n",
    "\n",
    "print(f\"Number of genes with either HIGH or MODERATE variants: {len(gene_SNP_annotation.keys())}\\n\")\n",
    "\n",
    "print(f\"Number of genes with HIGH or MODERATE variants with each annotation category: {annotation_counts}\\n\")\n",
    "\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "\tannotation_list = \",\".join(list(annotations))\n",
    "\tprint(f\"Gene {gene} annotations: {annotation_list}\")\n",
    "\n",
    "disruptive = {\"stop_gained\", \"frameshift_variant\", \"start_lost\"}\n",
    "\n",
    "if annotation_input:\n",
    "\tgene_annotation = pd.read_csv(annotation_input, delimiter=\"\\t\",engine=\"c\",index_col=0)\n",
    "\n",
    "\n",
    "gene_annotation.head()\n",
    "\n",
    "first_experiment_genes = set()\n",
    "\n",
    "print(\"\\nGenes with most impacting variants:\\n\")\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "    if len(annotations & disruptive) >= 1:\n",
    "        annotation_list = \", \".join(list(annotations))\n",
    "        pannzer = gene_annotation.at[gene,\"Pannzer annotation\"]\n",
    "        try:\n",
    "            pannzer1 = pannzer.split(\";\")[0]#Just use first in list otherwise too many\n",
    "        except:#If pannzer object is nan\n",
    "            pannzer1 = pannzer\n",
    "        print(f\"Gene {gene} ({pannzer1}, expressed?: {gene in expressed_genes}): {annotation_list} \\n\")\n",
    "        first_experiment_genes.add(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes with each impact category: {'MODIFIER': 131, 'LOW': 68, 'MODERATE': 52, 'HIGH': 11}\n",
      "Number of genes with either HIGH or MODERATE variants: 55\n",
      "Number of genes with HIGH or MODERATE variants with each annotation category: {'missense_variant': 52, 'splice_acceptor_variant&intron_variant': 3, 'missense_variant&splice_region_variant': 4, 'stop_gained': 1, 'frameshift_variant': 4, 'splice_donor_variant&intron_variant': 2, 'stop_lost': 2, 'disruptive_inframe_insertion': 1}\n",
      "Gene BaRT2v18chr3HG122610 annotations: missense_variant,splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG122620 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122640 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122680 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122690 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122720 annotations: missense_variant,missense_variant&splice_region_variant\n",
      "Gene BaRT2v18chr3HG122730 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122740 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122750 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122760 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122770 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122790 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122800 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122810 annotations: missense_variant,stop_gained,missense_variant&splice_region_variant\n",
      "Gene BaRT2v18chr3HG122810-BaRT2v18chr3HG122820 annotations: missense_variant,missense_variant&splice_region_variant\n",
      "Gene BaRT2v18chr3HG122830 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG122840 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122860 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG122870 annotations: splice_donor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG122880 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122890 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122900 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122920 annotations: missense_variant,missense_variant&splice_region_variant\n",
      "Gene BaRT2v18chr3HG122970 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122990 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123000 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123020 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123010 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123050 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123040 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123060 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123080 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123090 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123110 annotations: missense_variant,stop_lost,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123120 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123140 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123160 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123180 annotations: stop_lost\n",
      "Gene BaRT2v18chr3HG123220 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123250 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123260 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123280 annotations: disruptive_inframe_insertion,missense_variant\n",
      "Gene BaRT2v18chr3HG123380 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123400 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123410 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123480 annotations: missense_variant,splice_donor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123550 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123600 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123610 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123630 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123650 annotations: missense_variant,splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123660 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123700 annotations: splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123710 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123780 annotations: missense_variant\n",
      "\n",
      "Genes with most impacting variants:\n",
      "\n",
      "Gene BaRT2v18chr3HG122810 (Glycosyltransferase, ALG3, expressed?: True): missense_variant, stop_gained, missense_variant&splice_region_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG122830 (Signal recognition particle SEC65 subunit, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG122860 (nan, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123110 (nan, expressed?: True): missense_variant, stop_lost, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123140 (Xyloglucan endotransglucosylase/hydrolase, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Which genes with big variants are present in both experiments?\n",
      "\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'first_experiment_genes' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-1f6855d68936>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Which genes with big variants are present in both experiments?\\n\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mgene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mannotations\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgene_SNP_annotation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 85\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mgene\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfirst_experiment_genes\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     86\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mannotations\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0mdisruptive\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     87\u001b[0m             \u001b[0mannotation_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\", \"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mannotations\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'first_experiment_genes' is not defined"
     ]
    }
   ],
   "source": [
    "input_file = input_file2\n",
    "\n",
    "annotation_input = \"/cluster/db/mecoulter/BaRT2v18/BaRT_2_18_annotation_genes.txt\"\n",
    "\n",
    "expression_info = \"/cluster/db/mecoulter/RNAseq2/RNAquant_analysis_result/expressed_genes_both_adj_p_values.csv\"\n",
    "\n",
    "\n",
    "gene_annotation = pd.read_csv(expression_info,engine=\"c\",index_col=0)\n",
    "\n",
    "expressed_genes = set(gene_annotation.index)\n",
    "\n",
    "#output_file = sys.argv[2]\n",
    "\n",
    "impact = {\"MODIFIER\":0, \"LOW\": 0, \"MODERATE\": 0, \"HIGH\": 0 }\n",
    "\n",
    "gene_impact = {} #Per gene impact annotation\n",
    "gene_SNP_annotation = {} #Per gene SNP annotation, HIGH or MODERATE only\n",
    "\n",
    "annotation_counts = {}\n",
    "\n",
    "for line in open(input_file):\n",
    "\tif line.startswith(\"#\"):\n",
    "\t\tcontinue\n",
    "\tfields = line.rstrip().split(\"\\t\")\n",
    "\tINFO = fields[7]\n",
    "\tSNP_position = fields[1]\n",
    "\tannotations = INFO.split(\"ANN=\")[1].split(\"=\")[0].split(\",\")#List of all possible annotations\n",
    "\tfor annotation in annotations:\n",
    "\t\ttry:\n",
    "\t\t\tallele, p_annotation, putative_impact, position, gene, feature_type, transcript, transcript_type  = annotation.split(\"|\")[:8]\n",
    "\t\texcept ValueError:\n",
    "\t\t\tprint(f\"Error at position {SNP_position}! Annotation is {annotation}\")\n",
    "\t\t\tsys.exit()\n",
    "\t\tgene_impact.setdefault(gene,set()).add(putative_impact)\n",
    "\t\tif putative_impact == \"HIGH\" or putative_impact == \"MODERATE\":\n",
    "\t\t\tgene_SNP_annotation.setdefault(gene,set()).add(p_annotation)\n",
    "\n",
    "for gene, impacts in gene_impact.items():\n",
    "\tfor putative_impact in impacts:\n",
    "\t\timpact[putative_impact] += 1\n",
    "\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "\tfor annotation in annotations:\n",
    "\t\ttry:\n",
    "\t\t\tannotation_counts[annotation] += 1\n",
    "\t\texcept KeyError:\n",
    "\t\t\tannotation_counts[annotation] = 1\n",
    "\n",
    "print(f\"Number of genes with each impact category: {impact}\")#One gene can have multiple categories\n",
    "\n",
    "print(f\"Number of genes with either HIGH or MODERATE variants: {len(gene_SNP_annotation.keys())}\")\n",
    "\n",
    "print(f\"Number of genes with HIGH or MODERATE variants with each annotation category: {annotation_counts}\")\n",
    "\n",
    "\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "\tannotation_list = \",\".join(list(annotations))\n",
    "\tprint(f\"Gene {gene} annotations: {annotation_list}\")\n",
    "    \n",
    "\n",
    "disruptive = {\"stop_gained\", \"frameshift_variant\", \"start_lost\"}\n",
    "\n",
    "if annotation_input:\n",
    "\tgene_annotation = pd.read_csv(annotation_input, delimiter=\"\\t\",engine=\"c\",index_col=0)\n",
    "\n",
    "\n",
    "gene_annotation.head()\n",
    "\n",
    "print(\"\\nGenes with most impacting variants:\\n\")\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "    if len(annotations & disruptive) >= 1:\n",
    "        annotation_list = \", \".join(list(annotations))\n",
    "        pannzer = gene_annotation.at[gene,\"Pannzer annotation\"]\n",
    "        try:\n",
    "            pannzer1 = pannzer.split(\";\")[0]#Just use first in list otherwise too many\n",
    "        except:#If pannzer object is nan\n",
    "            pannzer1 = pannzer\n",
    "        print(f\"Gene {gene} ({pannzer1}, expressed?: {gene in expressed_genes}): {annotation_list} \\n\")\n",
    "\n",
    "\n",
    "print(\"\\n\\n\")\n",
    "        \n",
    "print(\"Which genes with big variants are present in both experiments?\\n\")\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "    if gene in first_experiment_genes:\n",
    "        if len(annotations & disruptive) >= 1:\n",
    "            annotation_list = \", \".join(list(annotations))\n",
    "            pannzer = gene_annotation.at[gene,\"Pannzer annotation\"]\n",
    "            try:\n",
    "                pannzer1 = pannzer.split(\";\")[0]#Just use first in list otherwise too many\n",
    "            except:#If pannzer object is nan\n",
    "                pannzer1 = pannzer\n",
    "            print(f\"Gene {gene} ({pannzer1}, expressed?: {gene in expressed_genes}): {annotation_list} \\n\")\n",
    "\n",
    "#Create table of results:\n",
    "table = pd.DataFrame({\"Genes\":list(gene_SNP_annotation.keys()), \"Variant annotation\": list(gene_SNP_annotation.values())})\n",
    "\n",
    "table.to_csv(path_or_buf=\"/cluster/db/mecoulter/RNAseq2/high_impact_variant_annotation.csv\",header=True)\n",
    "\n",
    "\t\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
