{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare variant calls and annotations from two RNA-seq experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "545 unique variants in experiment two\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "#input_file1 = \"/cluster/db/mecoulter/CEMartinez_RNAseq-209915706/results/bcftools/3H_locus_filtered.anno.vcf\"\n",
    "input_file2 = \"/cluster/db/mecoulter/RNAseq2/bcftools/3H_locus_filtered.anno.vcf\"\n",
    "\n",
    "first_set = set()\n",
    "second_set = set()\n",
    "\n",
    "locus = [33181340, 36970860]#3H locus boundaries on 3H\n",
    "\"\"\"\n",
    "for line in open(input_file1):#Barke - Int19/Int56\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    fields = line.rstrip().split(\"\\t\")\n",
    "    SNP_position = fields[1]\n",
    "    ref = fields[3]\n",
    "    alt = fields[4]\n",
    "    variant_id = f\"{SNP_position}_{ref}_{alt}\"\n",
    "    first_set.add(variant_id)\n",
    "\n",
    "print(f\"{len(first_set)} unique variants in experiment one\")\n",
    "\"\"\"\n",
    "for line in open(input_file2):#Barke/Int_124_17 - Int_124_52\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    fields = line.rstrip().split(\"\\t\")\n",
    "    SNP_position = fields[1]\n",
    "    if int(SNP_position) < locus[0] or int(SNP_position) > locus[1]:\n",
    "        continue\n",
    "    ref = fields[3]\n",
    "    alt = fields[4]\n",
    "    variant_id = f\"{SNP_position}_{ref}_{alt}\"\n",
    "    second_set.add(variant_id)\n",
    "\n",
    "print(f\"{len(second_set)} unique variants in experiment two\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* What variants are shared between two experiments?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(f\"Number of shared variants: {len(first_set & second_set)}, proportion (compared to first set): {len(first_set & second_set)/len(first_set)} \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = input_file2\n",
    "\n",
    "annotation_input = \"/cluster/db/mecoulter/BaRT2v18/BaRT_2_18_annotation_genes.txt\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class SnpEffAnnotation:\n",
    "    def __init__(self, annotation):\n",
    "        try:\n",
    "            self.allele, self.p_annotation, self.putative_impact, self.position, self.gene, self.feature_type, self.transcript, self.transcript_type  = annotation.split(\"|\")[:8]\n",
    "        except ValueError:\n",
    "            print(f\"Error at position {SNP_position}! Annotation is {annotation}\")\n",
    "            sys.exit()\n",
    "    def __repr__(self):\n",
    "        return self.p_annotation\n",
    "\n",
    "class SnpEff:\n",
    "    gene_impact = {}\n",
    "    gene_SNP_annotation = {}\n",
    "    Instances = {}#Gene: {position1: [annotations], position2: [annotations]}\n",
    "    Instances_clean = {}\n",
    "    position_info = {}\n",
    "    dn_ds = {}\n",
    "    \n",
    "    def __init__(self, line):\n",
    "        self.fields = line.rstrip().split(\"\\t\")\n",
    "        self.INFO = self.fields[7]\n",
    "        self.SNP_position = self.fields[1]\n",
    "        self.reference_allele = self.fields[3]\n",
    "        self.alternative_allele = self.fields[4]\n",
    "        annotations = self.INFO.split(\"ANN=\")[1].split(\"=\")[0].split(\",\")#List of all possible annotations\n",
    "        self.annotations = []\n",
    "        for a in annotations:\n",
    "            annotation = SnpEffAnnotation(a)\n",
    "            self.annotations.append(annotation)\n",
    "            self.gene_impact.setdefault(annotation.gene,[]).append(putative_impact)  \n",
    "            \n",
    "            if annotation.putative_impact == \"HIGH\" or annotation.putative_impact == \"MODERATE\":\n",
    "                gene_SNP_annotation.setdefault(annotation.gene,[]).append(f\"{annotation.p_annotation} : {self.SNP_position}\")\n",
    "                \n",
    "            try:\n",
    "                position_annotations = self.Instances[annotation.gene]\n",
    "                position_annotations.setdefault(self.SNP_position, []).append(annotation)\n",
    "                self.Instances[annotation.gene] = position_annotations\n",
    "            except KeyError:\n",
    "                positions_annotations = {}\n",
    "                positions_annotations[self.SNP_position] = [annotation]\n",
    "                self.Instances[annotation.gene] = positions_annotations\n",
    "                \n",
    "        self.position_info[self.SNP_position] = self\n",
    "                \n",
    "    def __getitem__(self, item):\n",
    "        if self.Instances_clean:\n",
    "            return self.Instances_clean[item]\n",
    "        else:\n",
    "            return self.Instances[item]\n",
    "    \n",
    "    def clean_up_annotations(self):\n",
    "        \"\"\"Many positions have multiple annotations.\n",
    "        This is because of multiple transcripts. Most transcripts will be similar, \n",
    "        so will effectively have the same annotation. So remove duplicates. Will\n",
    "        also remove annotations not of interest e.g upstream/ downstream\"\"\"\n",
    "        \n",
    "        not_interesting = {\"downstream_gene_variant\", \"upstream_gene_variant\", \\\n",
    "        \"intergenic_region\", \"intron_variant\"}\n",
    "        \n",
    "        for gene, position_annotation in self.Instances.items():\n",
    "            position_annotation_clean = {}\n",
    "            for position, annotations in position_annotation.items():\n",
    "                annotations_cleaned = []\n",
    "                all_p_annotations = set()\n",
    "                for annotation in annotations:\n",
    "                    if annotation.p_annotation not in all_p_annotations and annotation.p_annotation not in not_interesting:\n",
    "                        #if annotation.putative_impact == \"HIGH\" or annotation.putative_impact == \"MODERATE\":\n",
    "                        annotations_cleaned.append(annotation)\n",
    "                        all_p_annotations.add(annotation.p_annotation)\n",
    "                if annotations_cleaned:#Only keep high/moderate impact snps\n",
    "                    position_annotation_clean[position] = annotations_cleaned\n",
    "            self.Instances_clean[gene] = position_annotation_clean\n",
    "    def __repr__(self):\n",
    "        string = \"\"\n",
    "        print_dict = self.Instances_clean if self.Instances_clean else self.Instances\n",
    "        \n",
    "        for gene, position_annotation in print_dict.items():\n",
    "            string += f\"{gene}: \\n\"\n",
    "            for position, annotations in position_annotation.items():\n",
    "                string += f\"{position} : {','.join([str(annotation) for annotation in annotations])}\\n\"\n",
    "            string += \"\\n\"  \n",
    "        return string\n",
    "    \n",
    "    def __len__(self):\n",
    "        return(len(self.Instances.keys()))\n",
    "    \n",
    "    def write_tsv(self, outfile, gene_annotation):\n",
    "        \"\"\"Write a .csv file with results of cleaned up annotation. \n",
    "        Gene: Position: Annotation\"\"\"\n",
    "        with open(outfile, \"w\") as out:\n",
    "            out.write(\"Gene\\tPosition\\tAnnotation\\tref allele\\talt allele\\tPanzzer Annotation\\tdn/ds ratio\\n\")\n",
    "            for gene, position_annotation in self.Instances_clean.items():\n",
    "                \n",
    "                index = 0\n",
    "                \n",
    "                try:\n",
    "                    panz_annotation = gene_annotation.at[gene,\"Pannzer annotation\"]\n",
    "                except KeyError:\n",
    "                    print(f\"KeyError for gene {gene}\")\n",
    "                    panz_annotation = \"Not found\"\n",
    "                    \n",
    "                try:\n",
    "                    dn_ds = self.dn_ds[gene]\n",
    "                except KeyError:\n",
    "                    dn_ds = \"\"\n",
    "                    \n",
    "                for position, annotations in position_annotation.items():\n",
    "                    \n",
    "                    ref = self.position_info[position].reference_allele\n",
    "                    alt = self.position_info[position].alternative_allele\n",
    "                    \n",
    "                    out.write(f\"{gene}\\t\") if not index else out.write(f\"\\t\")\n",
    "                    \n",
    "                    out.write(f\"{position}\\t{','.join([str(annotation) for annotation in annotations])}\\t\\\n",
    "                    {ref}\\t{alt}\\t\")#\n",
    "                    \n",
    "                    #Only write annotation once\n",
    "                    out.write(f\"{panz_annotation}\\t{dn_ds}\\n\") if not index else out.write(\"\\t\\n\")\n",
    "                    \n",
    "                    index += 1\n",
    "            \n",
    "    def get_dn_ds(self):\n",
    "        \n",
    "        \"\"\"Calculates dn/ds ratio for each gene where possible\n",
    "        missense_variant = n, synonymous_variant\"\"\"\n",
    "        for gene, position_annotation in self.Instances_clean.items():\n",
    "            n = []\n",
    "            s = []\n",
    "            for position, annotations in position_annotation.items():\n",
    "                SNP_anotations = set([str(annotation) for annotation in annotations])\n",
    "                if \"missense_variant\" in SNP_anotations and \"synonymous_variant\" in SNP_anotations:\n",
    "                    continue#A case where SNP is annotated as both, these cannot be used\n",
    "                n += [annotation for annotation in SNP_anotations if annotation == \"missense_variant\"]\n",
    "                s += [annotation for annotation in SNP_anotations if annotation == \"synonymous_variant\"]\n",
    "            if n and s:\n",
    "                self.dn_ds[gene] = len(n)/len(s)\n",
    "            else:\n",
    "                self.dn_ds[gene] = \"NA\"\n",
    "            \n",
    "                \n",
    "                    \n",
    "                \n",
    "            \n",
    "        \n",
    "        \n",
    "        \n",
    "            \n",
    "\n",
    "\n",
    "gene_annotation = pd.read_csv(annotation_input,engine=\"c\",index_col=0, sep=\"\\t\")\n",
    "\n",
    "#expressed_genes = set(gene_annotation.index)\n",
    "\n",
    "\n",
    "\n",
    "impact = {\"MODIFIER\":0, \"LOW\": 0, \"MODERATE\": 0, \"HIGH\": 0 }\n",
    "locus = [33181340, 36970860]#3H locus boundaries on 3H\n",
    "\n",
    "for line in open(input_file):\n",
    "    #print(line)\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    #print(line)\n",
    "    position = int(line.rstrip().split(\"\\t\")[1])\n",
    "    if position < locus[0] or position > locus[1]:#ONly include snps in 3H locus\n",
    "        continue\n",
    "    snpeff = SnpEff(line)\n",
    "\n",
    "snpeff.clean_up_annotations()\n",
    "\n",
    "#print(snpeff)\n",
    "#print(f\"Length of Instances is {len(snpeff)}\")\n",
    "\n",
    "snpeff.get_dn_ds()#Get dn/ds ratio for each gene\n",
    "\n",
    "snpeff.write_tsv(\"/cluster/db/mecoulter/RNAseq2/3H_locus_snpeff_gene_annotation.tab\", gene_annotation)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_info = \"/cluster/db/mecoulter/RNAseq2/RNAquant_analysis_result/candidate_gene_complete_info.csv\"\n",
    "\n",
    "snp_annotation = pd.read_csv(\"/cluster/db/mecoulter/RNAseq2/3H_locus_snpeff_gene_annotation.tab\",\\\n",
    " engine=\"c\", sep=\"\\t\")\n",
    "\n",
    "expression = pd.read_csv(expression_info, engine=\"c\", index_col=1)\n",
    "\n",
    "expressed_genes = set(expression.index)\n",
    "#snp_annotation.head()\n",
    "\n",
    "snp_annotation_expressed = snp_annotation\n",
    "\n",
    "drop = False\n",
    "for i in snp_annotation.index:\n",
    "    if str(snp_annotation.at[i,\"Gene\"]) != \"nan\":\n",
    "        if snp_annotation.at[i,\"Gene\"] not in expressed_genes:\n",
    "            snp_annotation_expressed = snp_annotation_expressed.drop([i])\n",
    "            drop = True\n",
    "        else:\n",
    "            drop = False\n",
    "    else:\n",
    "        if drop:\n",
    "            snp_annotation_expressed = snp_annotation_expressed.drop([i])\n",
    "\n",
    "snp_annotation_expressed.to_csv(\"/cluster/db/mecoulter/RNAseq2/3H_locus_snpeff_gene_annotation_expressed.tab\", sep=\"\\t\")\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Position</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>ref allele</th>\n",
       "      <th>alt allele</th>\n",
       "      <th>Panzzer Annotation</th>\n",
       "      <th>dn/ds ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BaRT2v18chr3HG122970</td>\n",
       "      <td>33294871</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pentatricopeptide repeat-containing protein</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>33294891</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>33294929</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>33295153</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>33295177</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene  Position           Annotation  \\\n",
       "9   BaRT2v18chr3HG122970  33294871  3_prime_UTR_variant   \n",
       "10                   NaN  33294891  3_prime_UTR_variant   \n",
       "11                   NaN  33294929  3_prime_UTR_variant   \n",
       "12                   NaN  33295153  3_prime_UTR_variant   \n",
       "13                   NaN  33295177  3_prime_UTR_variant   \n",
       "\n",
       "                ref allele alt allele  \\\n",
       "9                        C          T   \n",
       "10                       A          G   \n",
       "11                       C          G   \n",
       "12                       T          C   \n",
       "13                      CT          C   \n",
       "\n",
       "                             Panzzer Annotation  dn/ds ratio  \n",
       "9   Pentatricopeptide repeat-containing protein          2.0  \n",
       "10                                          NaN          NaN  \n",
       "11                                          NaN          NaN  \n",
       "12                                          NaN          NaN  \n",
       "13                                          NaN          NaN  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_annotation_expressed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>raw_p</th>\n",
       "      <th>adj_p</th>\n",
       "      <th>lgfc</th>\n",
       "      <th>function.</th>\n",
       "      <th>GO.IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BaRT2v18chr3HG122970</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>33294773</td>\n",
       "      <td>33299957</td>\n",
       "      <td>-</td>\n",
       "      <td>0.273116</td>\n",
       "      <td>0.940247</td>\n",
       "      <td>-0.146344</td>\n",
       "      <td>Pentatricopeptide repeat-containing protein</td>\n",
       "      <td>GO:0033108;GO:0044085;GO:0006139;GO:0006996;GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BaRT2v18chr3HG122980</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>33300256</td>\n",
       "      <td>33301691</td>\n",
       "      <td>-</td>\n",
       "      <td>0.476399</td>\n",
       "      <td>0.950981</td>\n",
       "      <td>0.165490</td>\n",
       "      <td>Usp domain-containing protein</td>\n",
       "      <td>GO:0097159;GO:0009506;GO:0006457;GO:0043229;GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BaRT2v18chr3HG122990</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>33302620</td>\n",
       "      <td>33306323</td>\n",
       "      <td>-</td>\n",
       "      <td>0.641557</td>\n",
       "      <td>0.967256</td>\n",
       "      <td>0.051601</td>\n",
       "      <td>Undefined</td>\n",
       "      <td>GO:0003824;GO:0031224;GO:0110165;GO:0016021;GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BaRT2v18chr3HG123000</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>33574180</td>\n",
       "      <td>33580981</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.001670</td>\n",
       "      <td>-0.549154</td>\n",
       "      <td>SUPPRESSOR OF ABI3-5 isoform X1;OCRE domain-co...</td>\n",
       "      <td>GO:0097159;GO:0006139;GO:0003723;GO:0043229;GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BaRT2v18chr3HG123020</td>\n",
       "      <td>chr3H</td>\n",
       "      <td>33585286</td>\n",
       "      <td>33589376</td>\n",
       "      <td>+</td>\n",
       "      <td>0.370263</td>\n",
       "      <td>0.940247</td>\n",
       "      <td>-0.223143</td>\n",
       "      <td>Receptor protein kinase, putative</td>\n",
       "      <td>GO:0097159;GO:0036094;GO:0030554;GO:0140096;GO...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene Chromosome     Start       End Strand     raw_p  \\\n",
       "1  BaRT2v18chr3HG122970      chr3H  33294773  33299957      -  0.273116   \n",
       "2  BaRT2v18chr3HG122980      chr3H  33300256  33301691      -  0.476399   \n",
       "3  BaRT2v18chr3HG122990      chr3H  33302620  33306323      -  0.641557   \n",
       "4  BaRT2v18chr3HG123000      chr3H  33574180  33580981      +  0.000006   \n",
       "5  BaRT2v18chr3HG123020      chr3H  33585286  33589376      +  0.370263   \n",
       "\n",
       "      adj_p      lgfc                                          function.  \\\n",
       "1  0.940247 -0.146344        Pentatricopeptide repeat-containing protein   \n",
       "2  0.950981  0.165490                      Usp domain-containing protein   \n",
       "3  0.967256  0.051601                                          Undefined   \n",
       "4  0.001670 -0.549154  SUPPRESSOR OF ABI3-5 isoform X1;OCRE domain-co...   \n",
       "5  0.940247 -0.223143                  Receptor protein kinase, putative   \n",
       "\n",
       "                                              GO.IDs  \n",
       "1  GO:0033108;GO:0044085;GO:0006139;GO:0006996;GO...  \n",
       "2  GO:0097159;GO:0009506;GO:0006457;GO:0043229;GO...  \n",
       "3  GO:0003824;GO:0031224;GO:0110165;GO:0016021;GO...  \n",
       "4  GO:0097159;GO:0006139;GO:0003723;GO:0043229;GO...  \n",
       "5  GO:0097159;GO:0036094;GO:0030554;GO:0140096;GO...  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['BaRT2v18chr1HG000020', 'BaRT2v18chr1HG000040', 'BaRT2v18chr1HG000050',\n",
       "       'BaRT2v18chr1HG000060', 'BaRT2v18chr1HG000070'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expressed_genes[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
