{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare variant calls and annotations from two RNA-seq experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/cluster/db/mecoulter/CEMartinez_RNAseq-209915706/results/bcftools/3H_locus_filtered.anno.vcf'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-7cd38c0f031e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0msecond_set\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_file1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;31m#Barke - Int19/Int56\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"#\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/cluster/db/mecoulter/CEMartinez_RNAseq-209915706/results/bcftools/3H_locus_filtered.anno.vcf'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "input_file1 = \"/cluster/db/mecoulter/CEMartinez_RNAseq-209915706/results/bcftools/3H_locus_filtered.anno.vcf\"\n",
    "input_file2 = \"/cluster/db/mecoulter/RNAseq2/bcftools/3H_locus_filtered.anno.vcf\"\n",
    "\n",
    "first_set = set()\n",
    "second_set = set()\n",
    "\n",
    "for line in open(input_file1):#Barke - Int19/Int56\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    fields = line.rstrip().split(\"\\t\")\n",
    "    SNP_position = fields[1]\n",
    "    ref = fields[3]\n",
    "    alt = fields[4]\n",
    "    variant_id = f\"{SNP_position}_{ref}_{alt}\"\n",
    "    first_set.add(variant_id)\n",
    "\n",
    "print(f\"{len(first_set)} unique variants in experiment one\")\n",
    "    \n",
    "for line in open(input_file2):#Barke/Int_124_17 - Int_124_52\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    fields = line.rstrip().split(\"\\t\")\n",
    "    SNP_position = fields[1]\n",
    "    ref = fields[3]\n",
    "    alt = fields[4]\n",
    "    variant_id = f\"{SNP_position}_{ref}_{alt}\"\n",
    "    second_set.add(variant_id)\n",
    "\n",
    "print(f\"{len(second_set)} unique variants in experiment two\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* What variants are shared between two experiments?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of shared variants: 407, proportion (compared to first set): 0.2600638977635783 \n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of shared variants: {len(first_set & second_set)}, proportion (compared to first set): {len(first_set & second_set)/len(first_set)} \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of expressed genes in RNAseq dataset: 19802\n",
      "\n",
      "Number of genes with each impact category: {'MODIFIER': 145, 'LOW': 72, 'MODERATE': 72, 'HIGH': 17}\n",
      "\n",
      "Number of genes with either HIGH or MODERATE variants: 74\n",
      "\n",
      "Number of genes with HIGH or MODERATE variants with each annotation category: {'missense_variant': 72, 'frameshift_variant': 9, 'splice_donor_variant&intron_variant': 2, 'disruptive_inframe_deletion': 4, 'stop_gained': 1, 'missense_variant&splice_region_variant': 3, 'splice_acceptor_variant&intron_variant': 4, 'disruptive_inframe_insertion': 1, 'conservative_inframe_insertion': 1, 'start_lost': 1, 'frameshift_variant&stop_gained': 1}\n",
      "\n",
      "Gene BaRT2v18chr3HG122620 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122640 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG122660 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122690 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122720 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122730 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122740 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122750 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122760 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122770 annotations: missense_variant,splice_donor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG122780 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122790 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122800 annotations: missense_variant,disruptive_inframe_deletion\n",
      "Gene BaRT2v18chr3HG122810 annotations: missense_variant,stop_gained,missense_variant&splice_region_variant\n",
      "Gene BaRT2v18chr3HG122810-BaRT2v18chr3HG122820 annotations: missense_variant,missense_variant&splice_region_variant\n",
      "Gene BaRT2v18chr3HG122820 annotations: missense_variant,disruptive_inframe_deletion\n",
      "Gene BaRT2v18chr3HG122830 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122860 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122870 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG122880 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122890 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG122910 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122920 annotations: missense_variant,missense_variant&splice_region_variant,splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG122930 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122970 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG122990 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123000 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123010 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123020 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123030 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123050 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123040 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123060 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123070 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123080 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123090 annotations: missense_variant,splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123100 annotations: splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123110 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123120 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123140 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123150 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123160 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123170 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123220 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123240 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123230 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123250 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123260 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123270 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123280 annotations: missense_variant,disruptive_inframe_insertion\n",
      "Gene BaRT2v18chr3HG123350 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123400 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123410 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123480 annotations: missense_variant,disruptive_inframe_deletion,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123540 annotations: missense_variant,splice_donor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123550 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123600 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123610 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123630 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123650 annotations: missense_variant,conservative_inframe_insertion\n",
      "Gene BaRT2v18chr3HG123660 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123700 annotations: splice_acceptor_variant&intron_variant\n",
      "Gene BaRT2v18chr3HG123710 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123730-BaRT2v18chr3HG123740 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123730 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123740 annotations: missense_variant,start_lost\n",
      "Gene BaRT2v18chr3HG123750 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123760 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123770 annotations: missense_variant,frameshift_variant\n",
      "Gene BaRT2v18chr3HG123780 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123790 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123800 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123810 annotations: missense_variant\n",
      "Gene BaRT2v18chr3HG123830 annotations: missense_variant,disruptive_inframe_deletion,frameshift_variant&stop_gained,frameshift_variant\n",
      "\n",
      "Genes with most impacting variants:\n",
      "\n",
      "Gene BaRT2v18chr3HG122640 (Rab-GAP TBC domain-containing protein, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG122810 (Glycosyltransferase, ALG3, expressed?: True): missense_variant, stop_gained, missense_variant&splice_region_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG122870 (Ankyrin-1, expressed?: False): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG122890 (nan, expressed?: False): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123110 (nan, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123400 (nan, expressed?: False): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123480 (Folate-sensitive fragile site protein Fra10Ac1, expressed?: True): missense_variant, disruptive_inframe_deletion, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123740 (TNase-like domain-containing protein, expressed?: False): missense_variant, start_lost \n",
      "\n",
      "Gene BaRT2v18chr3HG123760 (TNase-like domain-containing protein, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123770 (Cellulose synthase-like protein H1-like, expressed?: True): missense_variant, frameshift_variant \n",
      "\n",
      "Gene BaRT2v18chr3HG123830 (nan, expressed?: False): missense_variant, disruptive_inframe_deletion, frameshift_variant&stop_gained, frameshift_variant \n",
      "\n"
     ]
    }
   ],
   "source": [
    "input_file = input_file1 \n",
    "\n",
    "annotation_input = \"/cluster/db/mecoulter/BaRT2v18/BaRT_2_18_annotation_genes.txt\"\n",
    "\n",
    "expression_info = \"/cluster/db/mecoulter/RNAseq2/RNAquant_analysis_result/expressed_genes_both_adj_p_values.csv\"\n",
    "\n",
    "\n",
    "gene_annotation = pd.read_csv(expression_info,engine=\"c\",index_col=0)\n",
    "\n",
    "expressed_genes = set(gene_annotation.index)\n",
    "\n",
    "print(f\"Number of expressed genes in RNAseq dataset: {len(expressed_genes)}\\n\")#Sanity check\n",
    "\n",
    "\n",
    "\n",
    "impact = {\"MODIFIER\":0, \"LOW\": 0, \"MODERATE\": 0, \"HIGH\": 0 }\n",
    "\n",
    "gene_impact = {} #Per gene impact annotation\n",
    "gene_SNP_annotation = {} #Per gene SNP annotation, HIGH or MODERATE only\n",
    "\n",
    "annotation_counts = {}\n",
    "\n",
    "for line in open(input_file):\n",
    "\tif line.startswith(\"#\"):\n",
    "\t\tcontinue\n",
    "\tfields = line.rstrip().split(\"\\t\")\n",
    "\tINFO = fields[7]\n",
    "\tSNP_position = fields[1]\n",
    "\tannotations = INFO.split(\"ANN=\")[1].split(\"=\")[0].split(\",\")#List of all possible annotations\n",
    "\tfor annotation in annotations:\n",
    "\t\ttry:\n",
    "\t\t\tallele, p_annotation, putative_impact, position, gene, feature_type, transcript, transcript_type  = annotation.split(\"|\")[:8]\n",
    "\t\texcept ValueError:\n",
    "\t\t\tprint(f\"Error at position {SNP_position}! Annotation is {annotation}\")\n",
    "\t\t\tsys.exit()\n",
    "\t\tgene_impact.setdefault(gene,set()).add(putative_impact)\n",
    "\t\tif putative_impact == \"HIGH\" or putative_impact == \"MODERATE\":\n",
    "\t\t\tgene_SNP_annotation.setdefault(gene,set()).add(p_annotation)\n",
    "\n",
    "for gene, impacts in gene_impact.items():\n",
    "\tfor putative_impact in impacts:\n",
    "\t\timpact[putative_impact] += 1\n",
    "\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "\tfor annotation in annotations:\n",
    "\t\ttry:\n",
    "\t\t\tannotation_counts[annotation] += 1\n",
    "\t\texcept KeyError:\n",
    "\t\t\tannotation_counts[annotation] = 1\n",
    "\n",
    "print(f\"Number of genes with each impact category: {impact}\\n\")#One gene can have multiple categories\n",
    "\n",
    "print(f\"Number of genes with either HIGH or MODERATE variants: {len(gene_SNP_annotation.keys())}\\n\")\n",
    "\n",
    "print(f\"Number of genes with HIGH or MODERATE variants with each annotation category: {annotation_counts}\\n\")\n",
    "\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "\tannotation_list = \",\".join(list(annotations))\n",
    "\tprint(f\"Gene {gene} annotations: {annotation_list}\")\n",
    "\n",
    "disruptive = {\"stop_gained\", \"frameshift_variant\", \"start_lost\"}\n",
    "\n",
    "if annotation_input:\n",
    "\tgene_annotation = pd.read_csv(annotation_input, delimiter=\"\\t\",engine=\"c\",index_col=0)\n",
    "\n",
    "\n",
    "gene_annotation.head()\n",
    "\n",
    "first_experiment_genes = set()\n",
    "\n",
    "print(\"\\nGenes with most impacting variants:\\n\")\n",
    "for gene, annotations in gene_SNP_annotation.items():\n",
    "    if len(annotations & disruptive) >= 1:\n",
    "        annotation_list = \", \".join(list(annotations))\n",
    "        pannzer = gene_annotation.at[gene,\"Pannzer annotation\"]\n",
    "        try:\n",
    "            pannzer1 = pannzer.split(\";\")[0]#Just use first in list otherwise too many\n",
    "        except:#If pannzer object is nan\n",
    "            pannzer1 = pannzer\n",
    "        print(f\"Gene {gene} ({pannzer1}, expressed?: {gene in expressed_genes}): {annotation_list} \\n\")\n",
    "        first_experiment_genes.add(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BaRT2v18chr3HG122920: \n",
      "33237835 : 5_prime_UTR_variant\n",
      "33237849 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "33241825 : missense_variant,missense_variant&splice_region_variant,3_prime_UTR_variant\n",
      "33244171 : 3_prime_UTR_variant\n",
      "33244188 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG122930: \n",
      "33249270 : 5_prime_UTR_variant\n",
      "33249283 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "33250161 : synonymous_variant\n",
      "33250186 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG122940: \n",
      "\n",
      "BaRT2v18chr3HG122940-BaRT2v18chr3HG122950: \n",
      "\n",
      "BaRT2v18chr3HG122960: \n",
      "\n",
      "BaRT2v18chr3HG122950-BaRT2v18chr3HG122960: \n",
      "\n",
      "BaRT2v18chr3HG122970: \n",
      "33294871 : 3_prime_UTR_variant\n",
      "33294891 : 3_prime_UTR_variant\n",
      "33294929 : 3_prime_UTR_variant\n",
      "33295153 : 3_prime_UTR_variant\n",
      "33295177 : 3_prime_UTR_variant\n",
      "33295236 : 3_prime_UTR_variant\n",
      "33296098 : 3_prime_UTR_variant\n",
      "33296420 : 3_prime_UTR_variant\n",
      "33296572 : 3_prime_UTR_variant\n",
      "33297796 : 3_prime_UTR_variant\n",
      "33297999 : 3_prime_UTR_variant\n",
      "33298005 : 3_prime_UTR_variant\n",
      "33298006 : 3_prime_UTR_variant\n",
      "33298503 : missense_variant\n",
      "33298672 : missense_variant\n",
      "33298677 : missense_variant\n",
      "33298718 : synonymous_variant\n",
      "33298968 : missense_variant\n",
      "33299021 : synonymous_variant\n",
      "\n",
      "BaRT2v18chr3HG122980: \n",
      "33300272 : 3_prime_UTR_variant\n",
      "33300332 : 3_prime_UTR_variant\n",
      "33300334 : 3_prime_UTR_variant\n",
      "33300374 : 3_prime_UTR_variant\n",
      "33300655 : synonymous_variant\n",
      "33300944 : synonymous_variant\n",
      "33301668 : 5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG122990: \n",
      "33302658 : 3_prime_UTR_variant\n",
      "33302831 : synonymous_variant\n",
      "33305702 : missense_variant\n",
      "33305765 : missense_variant\n",
      "33305766 : synonymous_variant\n",
      "33305985 : missense_variant\n",
      "33305990 : synonymous_variant\n",
      "33306244 : synonymous_variant\n",
      "33306245 : missense_variant\n",
      "33306298 : 5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG122980-BaRT2v18chr3HG122990: \n",
      "\n",
      "BaRT2v18chr3HG122990-BaRT2v18chr3HG123000: \n",
      "\n",
      "BaRT2v18chr3HG123000: \n",
      "33574220 : 5_prime_UTR_variant\n",
      "33574254 : 5_prime_UTR_variant\n",
      "33574321 : synonymous_variant\n",
      "33576083 : 3_prime_UTR_variant\n",
      "33576106 : 3_prime_UTR_variant\n",
      "33577017 : missense_variant,3_prime_UTR_variant\n",
      "33577727 : synonymous_variant,3_prime_UTR_variant\n",
      "33577868 : synonymous_variant,3_prime_UTR_variant\n",
      "33578683 : missense_variant,3_prime_UTR_variant\n",
      "33580916 : splice_region_variant,3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123010: \n",
      "33583952 : 3_prime_UTR_variant\n",
      "33584530 : 3_prime_UTR_variant\n",
      "33584650 : 3_prime_UTR_variant\n",
      "33585221 : 3_prime_UTR_variant\n",
      "33585276 : 3_prime_UTR_variant\n",
      "33585379 : 3_prime_UTR_variant\n",
      "33585425 : 3_prime_UTR_variant\n",
      "33585446 : 3_prime_UTR_variant\n",
      "33585484 : 3_prime_UTR_variant\n",
      "33585917 : 3_prime_UTR_variant\n",
      "33586222 : 3_prime_UTR_variant\n",
      "33586252 : 3_prime_UTR_variant\n",
      "33587087 : missense_variant\n",
      "33587360 : 5_prime_UTR_variant\n",
      "33587459 : 5_prime_UTR_variant\n",
      "33588106 : 5_prime_UTR_variant\n",
      "33588226 : 5_prime_UTR_variant\n",
      "33589063 : 5_prime_UTR_variant\n",
      "33589215 : 5_prime_UTR_variant\n",
      "33589321 : 5_prime_UTR_variant\n",
      "33589322 : 5_prime_UTR_variant\n",
      "33589388 : 5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123020: \n",
      "33585379 : 5_prime_UTR_variant\n",
      "33585425 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "33585446 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "33585484 : 5_prime_UTR_variant\n",
      "33585917 : missense_variant\n",
      "33586222 : synonymous_variant\n",
      "33586252 : synonymous_variant\n",
      "33587087 : missense_variant\n",
      "33587360 : missense_variant,5_prime_UTR_variant\n",
      "33587459 : missense_variant\n",
      "33588106 : synonymous_variant\n",
      "33588226 : synonymous_variant\n",
      "33589063 : 3_prime_UTR_variant\n",
      "33589215 : 3_prime_UTR_variant\n",
      "33589321 : 3_prime_UTR_variant\n",
      "33589322 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123000-BaRT2v18chr3HG123010: \n",
      "\n",
      "BaRT2v18chr3HG123010-BaRT2v18chr3HG123030: \n",
      "\n",
      "BaRT2v18chr3HG123030: \n",
      "33648878 : 3_prime_UTR_variant\n",
      "33648926 : 3_prime_UTR_variant\n",
      "33649037 : 3_prime_UTR_variant\n",
      "33649080 : synonymous_variant,3_prime_UTR_variant\n",
      "33649146 : synonymous_variant,3_prime_UTR_variant\n",
      "33649575 : synonymous_variant,3_prime_UTR_variant\n",
      "33650866 : splice_region_variant&intron_variant\n",
      "33651066 : 3_prime_UTR_variant\n",
      "33652198 : synonymous_variant\n",
      "33652370 : 5_prime_UTR_variant\n",
      "33652580 : 5_prime_UTR_variant\n",
      "33652830 : 5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123040: \n",
      "33873335 : 5_prime_UTR_variant\n",
      "33873337 : 5_prime_UTR_variant\n",
      "33873343 : 5_prime_UTR_variant\n",
      "33873492 : missense_variant\n",
      "33873694 : synonymous_variant\n",
      "33873715 : missense_variant\n",
      "33873802 : missense_variant\n",
      "33873988 : synonymous_variant\n",
      "33875039 : 3_prime_UTR_variant\n",
      "33876153 : missense_variant,3_prime_UTR_variant\n",
      "33876216 : 3_prime_UTR_variant\n",
      "33876241 : 3_prime_UTR_variant\n",
      "33876282 : 3_prime_UTR_variant\n",
      "33876372 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123050: \n",
      "33873335 : synonymous_variant\n",
      "33873337 : missense_variant\n",
      "33873343 : missense_variant\n",
      "33873492 : missense_variant\n",
      "33873694 : missense_variant\n",
      "33873715 : missense_variant\n",
      "33873802 : missense_variant\n",
      "33873988 : missense_variant\n",
      "33875039 : 5_prime_UTR_variant\n",
      "33876153 : 5_prime_UTR_variant\n",
      "33876216 : 5_prime_UTR_variant\n",
      "33876241 : 5_prime_UTR_variant\n",
      "33876282 : 5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123030-BaRT2v18chr3HG123040: \n",
      "\n",
      "BaRT2v18chr3HG123040-BaRT2v18chr3HG123060: \n",
      "\n",
      "BaRT2v18chr3HG123060: \n",
      "33916152 : missense_variant\n",
      "33916709 : 3_prime_UTR_variant\n",
      "33916754 : 3_prime_UTR_variant\n",
      "33917015 : 3_prime_UTR_variant\n",
      "33918364 : missense_variant,3_prime_UTR_variant\n",
      "33918809 : synonymous_variant,3_prime_UTR_variant\n",
      "33918899 : missense_variant,3_prime_UTR_variant\n",
      "33919449 : 3_prime_UTR_variant\n",
      "33919700 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123070: \n",
      "\n",
      "BaRT2v18chr3HG123080: \n",
      "34347332 : splice_region_variant&intron_variant,5_prime_UTR_variant\n",
      "34347403 : 5_prime_UTR_variant\n",
      "34347404 : 5_prime_UTR_variant\n",
      "34347457 : 5_prime_UTR_variant\n",
      "34347624 : synonymous_variant\n",
      "34347967 : missense_variant\n",
      "34348002 : synonymous_variant\n",
      "34356216 : splice_region_variant&intron_variant\n",
      "34356256 : missense_variant\n",
      "34356257 : missense_variant\n",
      "34356694 : missense_variant\n",
      "34356712 : missense_variant\n",
      "34356830 : synonymous_variant\n",
      "34357135 : missense_variant,3_prime_UTR_variant\n",
      "34357465 : 3_prime_UTR_variant\n",
      "34357527 : 3_prime_UTR_variant\n",
      "34357544 : 3_prime_UTR_variant\n",
      "34357622 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123090: \n",
      "34538301 : missense_variant,5_prime_UTR_variant,3_prime_UTR_variant\n",
      "34538680 : 5_prime_UTR_variant,3_prime_UTR_variant\n",
      "34538707 : 5_prime_UTR_variant,3_prime_UTR_variant\n",
      "34538771 : 5_prime_UTR_variant,3_prime_UTR_variant\n",
      "34542090 : synonymous_variant,3_prime_UTR_variant\n",
      "34542313 : synonymous_variant,3_prime_UTR_variant\n",
      "34560902 : synonymous_variant,3_prime_UTR_variant\n",
      "34562922 : synonymous_variant,3_prime_UTR_variant\n",
      "34562964 : synonymous_variant,3_prime_UTR_variant\n",
      "34564241 : missense_variant,3_prime_UTR_variant\n",
      "34564453 : synonymous_variant,3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123100: \n",
      "34646132 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "34646150 : 5_prime_UTR_variant\n",
      "34646226 : 5_prime_UTR_variant\n",
      "34646299 : 5_prime_UTR_variant\n",
      "34647607 : synonymous_variant\n",
      "34647682 : synonymous_variant\n",
      "\n",
      "BaRT2v18chr3HG123110: \n",
      "34748456 : 5_prime_UTR_variant\n",
      "34748574 : 5_prime_UTR_variant\n",
      "34748648 : 5_prime_UTR_variant\n",
      "34749036 : 5_prime_UTR_variant\n",
      "34749044 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "34749487 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "34749525 : 5_prime_UTR_variant\n",
      "34749617 : 5_prime_UTR_variant\n",
      "34749683 : 5_prime_UTR_variant\n",
      "34749723 : 5_prime_UTR_variant\n",
      "34749786 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "34750071 : 5_prime_UTR_variant\n",
      "34750157 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "34750203 : 5_prime_UTR_variant\n",
      "34750217 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "34750342 : 5_prime_UTR_variant\n",
      "34750354 : 5_prime_UTR_variant\n",
      "34750477 : synonymous_variant\n",
      "34750705 : 3_prime_UTR_variant\n",
      "34750986 : 3_prime_UTR_variant\n",
      "34751060 : missense_variant,3_prime_UTR_variant\n",
      "34751106 : missense_variant,3_prime_UTR_variant\n",
      "34751148 : missense_variant,splice_region_variant&intron_variant,3_prime_UTR_variant\n",
      "34751252 : frameshift_variant,3_prime_UTR_variant\n",
      "34751342 : missense_variant,3_prime_UTR_variant\n",
      "34751416 : missense_variant,3_prime_UTR_variant\n",
      "34751473 : stop_lost,3_prime_UTR_variant\n",
      "34751502 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123110-BaRT2v18chr3HG123120: \n",
      "\n",
      "BaRT2v18chr3HG123120: \n",
      "35079862 : 5_prime_UTR_variant\n",
      "35079920 : 5_prime_UTR_variant\n",
      "35079950 : 5_prime_UTR_variant\n",
      "35080651 : splice_region_variant&synonymous_variant,synonymous_variant\n",
      "35082257 : synonymous_variant,3_prime_UTR_variant\n",
      "35082460 : missense_variant,3_prime_UTR_variant\n",
      "35082465 : missense_variant,3_prime_UTR_variant\n",
      "35082674 : synonymous_variant,3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123130: \n",
      "35119035 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "35120211 : 3_prime_UTR_variant\n",
      "35120238 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123140: \n",
      "35132007 : 3_prime_UTR_variant\n",
      "35132279 : 3_prime_UTR_variant\n",
      "35132380 : 3_prime_UTR_variant\n",
      "35132434 : 3_prime_UTR_variant\n",
      "35133408 : synonymous_variant\n",
      "35133421 : missense_variant\n",
      "35133481 : frameshift_variant\n",
      "35133509 : synonymous_variant\n",
      "\n",
      "BaRT2v18chr3HG123160: \n",
      "35151670 : missense_variant\n",
      "35152022 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123150: \n",
      "35151670 : synonymous_variant\n",
      "35152022 : synonymous_variant\n",
      "\n",
      "BaRT2v18chr3HG123150-BaRT2v18chr3HG123170: \n",
      "\n",
      "BaRT2v18chr3HG123180: \n",
      "35319569 : 3_prime_UTR_variant\n",
      "35319772 : stop_lost\n",
      "\n",
      "BaRT2v18chr3HG123190: \n",
      "35319772 : splice_region_variant&intron_variant,5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123200: \n",
      "\n",
      "BaRT2v18chr3HG123210: \n",
      "\n",
      "BaRT2v18chr3HG123180-BaRT2v18chr3HG123200: \n",
      "\n",
      "BaRT2v18chr3HG123220: \n",
      "35389119 : missense_variant\n",
      "35389331 : missense_variant\n",
      "\n",
      "BaRT2v18chr3HG123210-BaRT2v18chr3HG123220: \n",
      "\n",
      "BaRT2v18chr3HG123250: \n",
      "35489658 : missense_variant\n",
      "35489686 : missense_variant\n",
      "35489826 : missense_variant\n",
      "35490018 : 5_prime_UTR_variant\n",
      "35490067 : 5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123260: \n",
      "35766007 : missense_variant\n",
      "35766138 : synonymous_variant\n",
      "35766222 : synonymous_variant\n",
      "35766597 : synonymous_variant\n",
      "35766622 : synonymous_variant\n",
      "35766680 : missense_variant\n",
      "35766696 : synonymous_variant\n",
      "35766755 : missense_variant\n",
      "\n",
      "BaRT2v18chr3HG123270: \n",
      "36013467 : synonymous_variant\n",
      "36013908 : 5_prime_UTR_variant\n",
      "36014042 : 5_prime_UTR_variant\n",
      "36014079 : 5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123280: \n",
      "36013467 : 5_prime_UTR_variant\n",
      "36013908 : synonymous_variant\n",
      "36014042 : missense_variant\n",
      "36019106 : synonymous_variant\n",
      "36019155 : missense_variant\n",
      "36019356 : synonymous_variant\n",
      "36019370 : missense_variant\n",
      "36019463 : disruptive_inframe_insertion\n",
      "36019506 : synonymous_variant\n",
      "36019846 : missense_variant\n",
      "36020006 : missense_variant\n",
      "36020072 : missense_variant\n",
      "36020089 : missense_variant\n",
      "36020097 : synonymous_variant\n",
      "36020510 : 3_prime_UTR_variant\n",
      "36020537 : 3_prime_UTR_variant\n",
      "36020550 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123360: \n",
      "\n",
      "BaRT2v18chr3HG123350: \n",
      "36112349 : synonymous_variant\n",
      "36112858 : 3_prime_UTR_variant\n",
      "36112925 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123370: \n",
      "\n",
      "BaRT2v18chr3HG123380: \n",
      "36160756 : 3_prime_UTR_variant\n",
      "36161744 : synonymous_variant,3_prime_UTR_variant\n",
      "36161849 : synonymous_variant,3_prime_UTR_variant\n",
      "36161852 : synonymous_variant,3_prime_UTR_variant\n",
      "36162341 : missense_variant,3_prime_UTR_variant\n",
      "36163244 : 3_prime_UTR_variant\n",
      "36164336 : synonymous_variant,3_prime_UTR_variant\n",
      "36164764 : 3_prime_UTR_variant\n",
      "36165670 : missense_variant\n",
      "36166492 : 5_prime_UTR_variant\n",
      "36166773 : 5_prime_UTR_variant\n",
      "36166948 : 5_prime_UTR_variant\n",
      "36167173 : 5_prime_UTR_variant\n",
      "36167252 : 5_prime_UTR_variant\n",
      "36167404 : 5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123410: \n",
      "36310248 : missense_variant\n",
      "36310302 : synonymous_variant\n",
      "36310353 : synonymous_variant\n",
      "36310405 : synonymous_variant\n",
      "36310551 : synonymous_variant\n",
      "36311012 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123400: \n",
      "36310248 : missense_variant\n",
      "36310302 : missense_variant\n",
      "36310353 : missense_variant\n",
      "36310405 : synonymous_variant\n",
      "36310551 : missense_variant\n",
      "36311012 : missense_variant\n",
      "\n",
      "BaRT2v18chr3HG123390-BaRT2v18chr3HG123400: \n",
      "\n",
      "BaRT2v18chr3HG123410-BaRT2v18chr3HG123420: \n",
      "\n",
      "BaRT2v18chr3HG123420: \n",
      "36340259 : 3_prime_UTR_variant\n",
      "36340373 : 3_prime_UTR_variant\n",
      "36341289 : synonymous_variant,3_prime_UTR_variant\n",
      "36341534 : synonymous_variant,3_prime_UTR_variant\n",
      "36341645 : synonymous_variant,3_prime_UTR_variant\n",
      "36343629 : synonymous_variant\n",
      "36343638 : synonymous_variant\n",
      "36343792 : 5_prime_UTR_premature_start_codon_gain_variant,5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123430: \n",
      "36389236 : 3_prime_UTR_variant\n",
      "36389359 : 3_prime_UTR_variant\n",
      "36389403 : 3_prime_UTR_variant\n",
      "36389475 : 3_prime_UTR_variant\n",
      "36389539 : synonymous_variant,3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123420-BaRT2v18chr3HG123430: \n",
      "\n",
      "BaRT2v18chr3HG123440: \n",
      "36397547 : 3_prime_UTR_variant\n",
      "36397563 : 3_prime_UTR_variant\n",
      "36397586 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123440-BaRT2v18chr3HG123450: \n",
      "\n",
      "BaRT2v18chr3HG123480: \n",
      "36577111 : 3_prime_UTR_variant\n",
      "36577209 : splice_region_variant,3_prime_UTR_variant\n",
      "36577257 : 3_prime_UTR_variant\n",
      "36577301 : 3_prime_UTR_variant\n",
      "36577464 : 3_prime_UTR_variant\n",
      "36578235 : missense_variant,3_prime_UTR_variant\n",
      "36578975 : 3_prime_UTR_variant\n",
      "36580924 : 3_prime_UTR_variant\n",
      "36581027 : 3_prime_UTR_variant\n",
      "36581207 : 3_prime_UTR_variant\n",
      "36581689 : 3_prime_UTR_variant\n",
      "36581814 : 3_prime_UTR_variant\n",
      "36581923 : 3_prime_UTR_variant\n",
      "36582042 : missense_variant\n",
      "36582714 : splice_donor_variant&intron_variant\n",
      "36582792 : 5_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123470: \n",
      "\n",
      "BaRT2v18chr3HG123490: \n",
      "36857104 : 3_prime_UTR_variant\n",
      "36857114 : 3_prime_UTR_variant\n",
      "36859120 : 3_prime_UTR_variant\n",
      "\n",
      "BaRT2v18chr3HG123490-BaRT2v18chr3HG123500: \n",
      "\n",
      "BaRT2v18chr3HG123500: \n",
      "36890032 : synonymous_variant\n",
      "\n",
      "\n",
      "Length of Instances is 63\n",
      "KeyError for gene BaRT2v18chr3HG122940-BaRT2v18chr3HG122950\n",
      "KeyError for gene BaRT2v18chr3HG122950-BaRT2v18chr3HG122960\n",
      "KeyError for gene BaRT2v18chr3HG122980-BaRT2v18chr3HG122990\n",
      "KeyError for gene BaRT2v18chr3HG122990-BaRT2v18chr3HG123000\n",
      "KeyError for gene BaRT2v18chr3HG123000-BaRT2v18chr3HG123010\n",
      "KeyError for gene BaRT2v18chr3HG123010-BaRT2v18chr3HG123030\n",
      "KeyError for gene BaRT2v18chr3HG123030-BaRT2v18chr3HG123040\n",
      "KeyError for gene BaRT2v18chr3HG123040-BaRT2v18chr3HG123060\n",
      "KeyError for gene BaRT2v18chr3HG123110-BaRT2v18chr3HG123120\n",
      "KeyError for gene BaRT2v18chr3HG123150-BaRT2v18chr3HG123170\n",
      "KeyError for gene BaRT2v18chr3HG123180-BaRT2v18chr3HG123200\n",
      "KeyError for gene BaRT2v18chr3HG123210-BaRT2v18chr3HG123220\n",
      "KeyError for gene BaRT2v18chr3HG123390-BaRT2v18chr3HG123400\n",
      "KeyError for gene BaRT2v18chr3HG123410-BaRT2v18chr3HG123420\n",
      "KeyError for gene BaRT2v18chr3HG123420-BaRT2v18chr3HG123430\n",
      "KeyError for gene BaRT2v18chr3HG123440-BaRT2v18chr3HG123450\n",
      "KeyError for gene BaRT2v18chr3HG123490-BaRT2v18chr3HG123500\n"
     ]
    }
   ],
   "source": [
    "input_file = input_file2\n",
    "\n",
    "annotation_input = \"/cluster/db/mecoulter/BaRT2v18/BaRT_2_18_annotation_genes.txt\"\n",
    "\n",
    "expression_info = \"/cluster/db/mecoulter/RNAseq2/RNAquant_analysis_result/expressed_genes_both_adj_p_values.csv\"\n",
    "\n",
    "\n",
    "class SnpEffAnnotation:\n",
    "    def __init__(self, annotation):\n",
    "        try:\n",
    "            self.allele, self.p_annotation, self.putative_impact, self.position, self.gene, self.feature_type, self.transcript, self.transcript_type  = annotation.split(\"|\")[:8]\n",
    "        except ValueError:\n",
    "            print(f\"Error at position {SNP_position}! Annotation is {annotation}\")\n",
    "            sys.exit()\n",
    "    def __repr__(self):\n",
    "        return self.p_annotation\n",
    "\n",
    "class SnpEff:\n",
    "    gene_impact = {}\n",
    "    gene_SNP_annotation = {}\n",
    "    Instances = {}#Gene: {position1: [annotations], position2: [annotations]}\n",
    "    Instances_clean = {}\n",
    "    position_info = {}\n",
    "    \n",
    "    def __init__(self, line):\n",
    "        self.fields = line.rstrip().split(\"\\t\")\n",
    "        self.INFO = self.fields[7]\n",
    "        self.SNP_position = self.fields[1]\n",
    "        self.reference_allele = self.fields[3]\n",
    "        self.alternative_allele = self.fields[4]\n",
    "        annotations = self.INFO.split(\"ANN=\")[1].split(\"=\")[0].split(\",\")#List of all possible annotations\n",
    "        self.annotations = []\n",
    "        for a in annotations:\n",
    "            annotation = SnpEffAnnotation(a)\n",
    "            self.annotations.append(annotation)\n",
    "            self.gene_impact.setdefault(annotation.gene,[]).append(putative_impact)  \n",
    "            \n",
    "            if annotation.putative_impact == \"HIGH\" or annotation.putative_impact == \"MODERATE\":\n",
    "                gene_SNP_annotation.setdefault(annotation.gene,[]).append(f\"{annotation.p_annotation} : {self.SNP_position}\")\n",
    "                \n",
    "            try:\n",
    "                position_annotations = self.Instances[annotation.gene]\n",
    "                position_annotations.setdefault(self.SNP_position, []).append(annotation)\n",
    "                self.Instances[annotation.gene] = position_annotations\n",
    "            except KeyError:\n",
    "                positions_annotations = {}\n",
    "                positions_annotations[self.SNP_position] = [annotation]\n",
    "                self.Instances[annotation.gene] = positions_annotations\n",
    "                \n",
    "        self.position_info[self.SNP_position] = self\n",
    "                \n",
    "    def __getitem__(self, item):\n",
    "        if self.Instances_clean:\n",
    "            return self.Instances_clean[item]\n",
    "        else:\n",
    "            return self.Instances[item]\n",
    "    \n",
    "    def clean_up_annotations(self):\n",
    "        \"\"\"Many positions have multiple annotations.\n",
    "        This is because of multiple transcripts. Most transcripts will be similar, \n",
    "        so will effectively have the same annotation. So remove duplicates. Will\n",
    "        also remove annotations not of interest e.g upstream/ downstream\"\"\"\n",
    "        \n",
    "        not_interesting = {\"downstream_gene_variant\", \"upstream_gene_variant\", \\\n",
    "        \"intergenic_region\", \"intron_variant\"}\n",
    "        \n",
    "        for gene, position_annotation in self.Instances.items():\n",
    "            position_annotation_clean = {}\n",
    "            for position, annotations in position_annotation.items():\n",
    "                annotations_cleaned = []\n",
    "                all_p_annotations = set()\n",
    "                for annotation in annotations:\n",
    "                    if annotation.p_annotation not in all_p_annotations and annotation.p_annotation not in not_interesting:\n",
    "                        #if annotation.putative_impact == \"HIGH\" or annotation.putative_impact == \"MODERATE\":\n",
    "                        annotations_cleaned.append(annotation)\n",
    "                        all_p_annotations.add(annotation.p_annotation)\n",
    "                if annotations_cleaned:#Only keep high/moderate impact snps\n",
    "                    position_annotation_clean[position] = annotations_cleaned\n",
    "            self.Instances_clean[gene] = position_annotation_clean\n",
    "    def __repr__(self):\n",
    "        string = \"\"\n",
    "        print_dict = self.Instances_clean if self.Instances_clean else self.Instances\n",
    "        \n",
    "        for gene, position_annotation in print_dict.items():\n",
    "            string += f\"{gene}: \\n\"\n",
    "            for position, annotations in position_annotation.items():\n",
    "                string += f\"{position} : {','.join([str(annotation) for annotation in annotations])}\\n\"\n",
    "            string += \"\\n\"  \n",
    "        return string\n",
    "    \n",
    "    def __len__(self):\n",
    "        return(len(self.Instances.keys()))\n",
    "    \n",
    "    def write_tsv(self, outfile, gene_annotation):\n",
    "        \"\"\"Write a .csv file with results of cleaned up annotation. \n",
    "        Gene: Position: Annotation\"\"\"\n",
    "        with open(outfile, \"w\") as out:\n",
    "            out.write(\"Gene\\tPosition\\tAnnotation\\tref allele\\talt allele\\tPanzzer Annotation\\n\")\n",
    "            for gene, position_annotation in self.Instances_clean.items():\n",
    "                index = 0\n",
    "                try:\n",
    "                    panz_annotation = gene_annotation.at[gene,\"Pannzer annotation\"]\n",
    "                except KeyError:\n",
    "                    print(f\"KeyError for gene {gene}\")\n",
    "                    panz_annotation = \"Not found\"\n",
    "                for position, annotations in position_annotation.items():\n",
    "                    ref = self.position_info[position].reference_allele\n",
    "                    alt = self.position_info[position].alternative_allele\n",
    "                    out.write(f\"{gene}\\t{position}\\t{','.join([str(annotation) for annotation in annotations])}\\t\\\n",
    "                    {ref}\\t{alt}\\t\")#\n",
    "                    if not index:#Only write annotation once\n",
    "                        out.write(f\"{panz_annotation}\\n\")\n",
    "                        index += 1\n",
    "                    else:\n",
    "                        out.write(\"\\n\")\n",
    "        \n",
    "        \n",
    "        \n",
    "            \n",
    "\n",
    "\n",
    "gene_annotation = pd.read_csv(annotation_input,engine=\"c\",index_col=0, sep=\"\\t\")\n",
    "\n",
    "#expressed_genes = set(gene_annotation.index)\n",
    "\n",
    "\n",
    "\n",
    "impact = {\"MODIFIER\":0, \"LOW\": 0, \"MODERATE\": 0, \"HIGH\": 0 }\n",
    "locus = [33181340, 36970860]#3H locus boundaries on 3H\n",
    "\n",
    "for line in open(input_file):\n",
    "    #print(line)\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    #print(line)\n",
    "    position = int(line.rstrip().split(\"\\t\")[1])\n",
    "    if position < locus[0] or position > locus[1]:#ONly include snps in 3H locus\n",
    "        continue\n",
    "    snpeff = SnpEff(line)\n",
    "\n",
    "snpeff.clean_up_annotations()\n",
    "print(snpeff)\n",
    "print(f\"Length of Instances is {len(snpeff)}\")\n",
    "snpeff.write_tsv(\"/cluster/db/mecoulter/RNAseq2/3H_locus_snpeff_gene_annotation.tab\", gene_annotation)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HID19.1-Barke.1</th>\n",
       "      <th>HID56.1-Barke.1</th>\n",
       "      <th>HEB_124_17.2-Barke.2</th>\n",
       "      <th>HEB_124_52.2-Barke.2</th>\n",
       "      <th>HEB_124_52.2-HEB_124_17.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr1HG000020</th>\n",
       "      <td>0.002923</td>\n",
       "      <td>0.178220</td>\n",
       "      <td>0.953696</td>\n",
       "      <td>0.176870</td>\n",
       "      <td>0.936994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr1HG000040</th>\n",
       "      <td>0.008994</td>\n",
       "      <td>0.006047</td>\n",
       "      <td>0.984281</td>\n",
       "      <td>0.879735</td>\n",
       "      <td>0.940247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr1HG000050</th>\n",
       "      <td>0.063700</td>\n",
       "      <td>0.937603</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.953811</td>\n",
       "      <td>0.976945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr1HG000060</th>\n",
       "      <td>0.015122</td>\n",
       "      <td>0.169527</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.838232</td>\n",
       "      <td>0.970221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BaRT2v18chr1HG000070</th>\n",
       "      <td>0.000618</td>\n",
       "      <td>0.005162</td>\n",
       "      <td>0.985872</td>\n",
       "      <td>0.528943</td>\n",
       "      <td>0.920581</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      HID19.1-Barke.1  HID56.1-Barke.1  HEB_124_17.2-Barke.2  \\\n",
       "BaRT2v18chr1HG000020         0.002923         0.178220              0.953696   \n",
       "BaRT2v18chr1HG000040         0.008994         0.006047              0.984281   \n",
       "BaRT2v18chr1HG000050         0.063700         0.937603              1.000000   \n",
       "BaRT2v18chr1HG000060         0.015122         0.169527              1.000000   \n",
       "BaRT2v18chr1HG000070         0.000618         0.005162              0.985872   \n",
       "\n",
       "                      HEB_124_52.2-Barke.2  HEB_124_52.2-HEB_124_17.2  \n",
       "BaRT2v18chr1HG000020              0.176870                   0.936994  \n",
       "BaRT2v18chr1HG000040              0.879735                   0.940247  \n",
       "BaRT2v18chr1HG000050              0.953811                   0.976945  \n",
       "BaRT2v18chr1HG000060              0.838232                   0.970221  \n",
       "BaRT2v18chr1HG000070              0.528943                   0.920581  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation = pd.read_csv(expression_info,engine=\"c\",index_col=0)\n",
    "gene_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
